<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.549">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>ENTMLGY 6703 - Jelmer’s lectures - Removed</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">ENTMLGY 6703 - Jelmer’s lectures</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html"> <i class="bi bi-house" role="img">
</i> 
<span class="menu-text"> </span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-week-3-hts" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Week 3: HTS</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-week-3-hts">    
        <li>
    <a class="dropdown-item" href="../lectures/sequencing.html">
 <span class="dropdown-text">Lecture: High-throughput Sequencing &amp; Genomes</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../labs/infrastructure.html">
 <span class="dropdown-text">Lab part 1: Computational Infrastructure</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../labs/data.html">
 <span class="dropdown-text">Lab part 2: Working with genome and HTS Data</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-week-4-rna-seq" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Week 4: RNA-seq</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-week-4-rna-seq">    
        <li>
    <a class="dropdown-item" href="../lectures/rnaseq.html">
 <span class="dropdown-text">Lecture: RNA-seq</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../labs/DE.html">
 <span class="dropdown-text">Lab: Differential expression analysis</span></a>
  </li>  
    </ul>
  </li>
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-other-pages" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Other Pages</span>
    </a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-other-pages">    
        <li>
    <a class="dropdown-item" href="../r_homework.html">
 <span class="dropdown-text">R Homework</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item">
    <a class="nav-link" href="../about.html"> 
<span class="menu-text">About</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label=""><i class="bi bi-github"></i></a>
      <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="quarto-navigation-tool-dropdown-0">
          <li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://github.com/jelmerp/ENTMLGY6703">
            Source Code
            </a>
          </li>
          <li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://github.com/jelmerp/ENTMLGY6703/issues/new">
            Report an Issue
            </a>
          </li>
      </ul>
    </div>
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#sanger-sequencing" id="toc-sanger-sequencing" class="nav-link active" data-scroll-target="#sanger-sequencing"><span class="header-section-number">1</span> Sanger sequencing</a></li>
  <li><a href="#rna-seq-projects" id="toc-rna-seq-projects" class="nav-link" data-scroll-target="#rna-seq-projects"><span class="header-section-number">2</span> RNA-seq projects</a></li>
  <li><a href="#x" id="toc-x" class="nav-link" data-scroll-target="#x"><span class="header-section-number">3</span> x</a></li>
  <li><a href="#x-1" id="toc-x-1" class="nav-link" data-scroll-target="#x-1"><span class="header-section-number">4</span> x</a></li>
  <li><a href="#microarrays" id="toc-microarrays" class="nav-link" data-scroll-target="#microarrays"><span class="header-section-number">5</span> Microarrays</a></li>
  <li><a href="#overview-of-all-steps-in-rna-seq" id="toc-overview-of-all-steps-in-rna-seq" class="nav-link" data-scroll-target="#overview-of-all-steps-in-rna-seq"><span class="header-section-number">6</span> Overview of all steps in RNA-seq</a></li>
  <li><a href="#other-ways-of-quantifying-gene-expression" id="toc-other-ways-of-quantifying-gene-expression" class="nav-link" data-scroll-target="#other-ways-of-quantifying-gene-expression"><span class="header-section-number">7</span> Other ways of quantifying gene expression</a></li>
  <li><a href="#some-other-types-of-rna-seq" id="toc-some-other-types-of-rna-seq" class="nav-link" data-scroll-target="#some-other-types-of-rna-seq"><span class="header-section-number">8</span> Some other types of RNA-seq</a></li>
  <li><a href="#rna-extraction" id="toc-rna-extraction" class="nav-link" data-scroll-target="#rna-extraction"><span class="header-section-number">9</span> RNA extraction</a></li>
  <li><a href="#rin-rna-integrity-number-score" id="toc-rin-rna-integrity-number-score" class="nav-link" data-scroll-target="#rin-rna-integrity-number-score"><span class="header-section-number">10</span> RIN (RNA Integrity Number) score</a></li>
  <li><a href="#library-prep-options" id="toc-library-prep-options" class="nav-link" data-scroll-target="#library-prep-options"><span class="header-section-number">11</span> Library prep options</a></li>
  <li><a href="#functional-enrichment-statistics" id="toc-functional-enrichment-statistics" class="nav-link" data-scroll-target="#functional-enrichment-statistics"><span class="header-section-number">12</span> Functional enrichment: statistics</a></li>
  <li><a href="#de-analysis-other-designs" id="toc-de-analysis-other-designs" class="nav-link" data-scroll-target="#de-analysis-other-designs"><span class="header-section-number">13</span> DE analysis: other designs</a></li>
  <li><a href="#normalization-by-gene-length" id="toc-normalization-by-gene-length" class="nav-link" data-scroll-target="#normalization-by-gene-length"><span class="header-section-number">14</span> Normalization by gene length</a></li>
  <li><a href="#alternative-splicing-isoforms" id="toc-alternative-splicing-isoforms" class="nav-link" data-scroll-target="#alternative-splicing-isoforms"><span class="header-section-number">15</span> Alternative splicing &amp; isoforms</a></li>
  <li><a href="#alternative-splicing-isoforms-1" id="toc-alternative-splicing-isoforms-1" class="nav-link" data-scroll-target="#alternative-splicing-isoforms-1"><span class="header-section-number">16</span> Alternative splicing &amp; isoforms</a></li>
  <li><a href="#what-if-i-dont-have-a-reference-genome" id="toc-what-if-i-dont-have-a-reference-genome" class="nav-link" data-scroll-target="#what-if-i-dont-have-a-reference-genome"><span class="header-section-number">17</span> What if I don’t have a reference genome?</a></li>
  <li><a href="#reads-to-counts-alignment-to-a-reference-genome" id="toc-reads-to-counts-alignment-to-a-reference-genome" class="nav-link" data-scroll-target="#reads-to-counts-alignment-to-a-reference-genome"><span class="header-section-number">18</span> Reads to counts: alignment to a reference genome</a></li>
  <li><a href="#reads-to-counts-alignment-qc" id="toc-reads-to-counts-alignment-qc" class="nav-link" data-scroll-target="#reads-to-counts-alignment-qc"><span class="header-section-number">19</span> Reads to counts: alignment QC</a></li>
  <li><a href="#reads-to-counts-alignment-qc-1" id="toc-reads-to-counts-alignment-qc-1" class="nav-link" data-scroll-target="#reads-to-counts-alignment-qc-1"><span class="header-section-number">20</span> Reads to counts: alignment QC</a></li>
  <li><a href="#reads-to-counts-read-pre-processing" id="toc-reads-to-counts-read-pre-processing" class="nav-link" data-scroll-target="#reads-to-counts-read-pre-processing"><span class="header-section-number">21</span> Reads to counts: Read pre-processing</a></li>
  <li><a href="#functional-enrichment" id="toc-functional-enrichment" class="nav-link" data-scroll-target="#functional-enrichment"><span class="header-section-number">22</span> Functional enrichment</a></li>
  <li><a href="#de-analysis-1-pairwise-comparison" id="toc-de-analysis-1-pairwise-comparison" class="nav-link" data-scroll-target="#de-analysis-1-pairwise-comparison"><span class="header-section-number">23</span> DE analysis: &gt;1 pairwise comparison</a></li>
  <li><a href="#side-note-dual-rna-seq" id="toc-side-note-dual-rna-seq" class="nav-link" data-scroll-target="#side-note-dual-rna-seq"><span class="header-section-number">24</span> Side note: Dual RNA-seq</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Removed</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="sanger-sequencing" class="level2" data-number="1">
<h2 data-number="1" class="anchored" data-anchor-id="sanger-sequencing"><span class="header-section-number">1</span> Sanger sequencing</h2>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<strong>Amplification</strong> of a target DNA fragment is usually done with PCR
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li><p>This means that you need to (approximately) <strong>know in advance short flanking sequences</strong> to the sequence of interest — primers for your PCR.</p></li>
<li><p><strong>Introns</strong> are good targets to sequence: variable sequences flanked by conserved sequences (exons) in which primers can be designed.</p></li>
</ul>
</div>
</div>
</section>
<section id="rna-seq-projects" class="level2" data-number="2">
<h2 data-number="2" class="anchored" data-anchor-id="rna-seq-projects"><span class="header-section-number">2</span> RNA-seq projects</h2>
<ul>
<li>Effects of <em>Diplodia</em> pathogen exposure Austrian pine (<em>Pinus nigra</em>) over time.</li>
<li>Aphid biotypes &amp; soybean cultivars under different flood conditions (Michel lab, Entomology)</li>
</ul>
</section>
<section id="x" class="level2" data-number="3">
<h2 data-number="3" class="anchored" data-anchor-id="x"><span class="header-section-number">3</span> x</h2>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img_sequencing/genome_sequencing1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
<figcaption>https://www.genome.gov/about-genomics/fact-sheets/Sequencing-Human-Genome-cost</figcaption>
</figure>
</div>
</section>
<section id="x-1" class="level2" data-number="4">
<h2 data-number="4" class="anchored" data-anchor-id="x-1"><span class="header-section-number">4</span> x</h2>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="https://www.genome.gov/about-genomics/fact-sheets/Sequencing-Human-Genome-cost"><img src="img_sequencing/genome_sequencing.jpeg" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%" alt="https://www.genome.gov/about-genomics/fact-sheets/Sequencing-Human-Genome-cost"></a></p>
<figcaption>https://www.genome.gov/about-genomics/fact-sheets/Sequencing-Human-Genome-cost</figcaption>
</figure>
</div>
</section>
<section id="microarrays" class="level2" data-number="5">
<h2 data-number="5" class="anchored" data-anchor-id="microarrays"><span class="header-section-number">5</span> Microarrays</h2>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="https://www.nature.com/articles/ng0199supp_33"><img src="img_rnaseq/microarray.webp" class="img-fluid figure-img" alt="Brown &amp; Botstein 1999 (www.nature.com/articles/ng0199supp_33)"></a></p>
<figcaption>Brown &amp; Botstein 1999 (www.nature.com/articles/ng0199supp_33)</figcaption>
</figure>
</div>
</section>
<section id="overview-of-all-steps-in-rna-seq" class="level2" data-number="6">
<h2 data-number="6" class="anchored" data-anchor-id="overview-of-all-steps-in-rna-seq"><span class="header-section-number">6</span> Overview of all steps in RNA-seq</h2>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="https://www.annualreviews.org/doi/10.1146/annurev-biodatasci-072018-021255"><img src="img_rnaseq/rnaseq_overview.jpeg" class="img-fluid quarto-figure quarto-figure-center figure-img" alt="Berge et al.&nbsp;2019 (www.annualreviews.org/doi/10.1146/annurev-biodatasci-072018-021255)"></a></p>
<figcaption>Berge et al.&nbsp;2019 (www.annualreviews.org/doi/10.1146/annurev-biodatasci-072018-021255)</figcaption>
</figure>
</div>
</section>
<section id="other-ways-of-quantifying-gene-expression" class="level2" data-number="7">
<h2 data-number="7" class="anchored" data-anchor-id="other-ways-of-quantifying-gene-expression"><span class="header-section-number">7</span> Other ways of quantifying gene expression</h2>
<p>RNA-seq is not the only method to quantify gene expression, but often the preferred one:</p>
<hr style="height:1pt; visibility:hidden;">
<ul>
<li><strong>Reverse-transcription qPCR</strong> (RT-qPCR)<br>
A low-throughput method that can quantify expression levels for one or a few genes at a time</li>
</ul>
<hr style="height:1pt; visibility:hidden;">
<p>. . .</p>
<ul>
<li><p><strong>Microarrays</strong> — hybridization of cDNA to preconstructed probes<br>
Microarrays were the transcriptomics method of choice before the advent of high-throughput sequencing, but disadvantages relative to RNA-seq include:</p>
<ul>
<li><p>Produce <strong>relative measurements</strong> only with a much <strong>smaller dynamic range</strong> than RNA-seq</p></li>
<li><p>Can only capture <strong>known features</strong>: cannot discover or examine novel features, and do not work for species without a (close) reference genome</p></li>
</ul></li>
</ul>
</section>
<section id="some-other-types-of-rna-seq" class="level2" data-number="8">
<h2 data-number="8" class="anchored" data-anchor-id="some-other-types-of-rna-seq"><span class="header-section-number">8</span> Some other types of RNA-seq</h2>
<ul>
<li><p><strong>Small RNA-seq</strong> for small noncoding RNAs such as microRNAs and Piwi-interacting RNAs</p></li>
<li><p><strong>Single-cell RNA-seq</strong> (scRNA-seq) — commonly used in cancer research</p></li>
<li><p><strong>Direct RNA sequencing</strong></p></li>
</ul>
<hr style="height:1pt; visibility:hidden;">
<p>:::</p>
</section>
<section id="rna-extraction" class="level2" data-number="9">
<h2 data-number="9" class="anchored" data-anchor-id="rna-extraction"><span class="header-section-number">9</span> RNA extraction</h2>
<p>The first step is to <strong>extract RNA from tissue samples</strong>. This is typically the only step you conduct yourself in the lab, with sequencing facilities taking care of downstream procedures.</p>
<p><br></p>
<p>Some considerations and recommendations:</p>
<p>. . .</p>
<ul>
<li><p>The tissue sample should be collected from an <strong>organism that is alive or recently deceased</strong> (matter of minutes), and the RNA should be extracted immediately.</p></li>
<li><p>With the appropriate buffer and temperate, an RNA extraction <em>can</em> be stored long-term.</p></li>
</ul>
<p>. . .</p>
<ul>
<li><p>Include a <strong>DNase treatment step</strong> to avoid DNA contamination.</p></li>
<li><p>Perform <strong>Quality Control (QC)</strong> on your extractions: quantity (e.g.&nbsp;with a Nanodrop or Qubit) and quality (degradation; e.g.&nbsp;with a gel). Sequencing facilities can additionally check the RIN score with a “Bioanalyzer”.</p></li>
</ul>
</section>
<section id="rin-rna-integrity-number-score" class="level2" data-number="10">
<h2 data-number="10" class="anchored" data-anchor-id="rin-rna-integrity-number-score"><span class="header-section-number">10</span> RIN (RNA Integrity Number) score</h2>
<div class="columns">
<div class="column" style="width:75%;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="https://www.novogene.com/eu-en/resources/blog/rna-integrity-number-rin-explained"><img src="img_rnaseq/RIN.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></a></p>
</figure>
</div>
<div class="legend2">
<p>https://www.novogene.com/eu-en/resources/blog/rna-integrity-number-rin-explained</p>
</div>
</div><div class="column" style="width:25%;">
<ul>
<li>The RIN score quantifies RNA quality: the lower the RIN score, the more degraded your RNA is likely to be.</li>
</ul>
<p><br></p>
<ul>
<li>It makes use the expectation of pronounced peaks corresponding to highly abundance, intact 18S and 28S rRNAs.</li>
</ul>
</div>
</div>
</section>
<section id="library-prep-options" class="level2" data-number="11">
<h2 data-number="11" class="anchored" data-anchor-id="library-prep-options"><span class="header-section-number">11</span> Library prep options</h2>
<p>Let’s consider the different options you have for RNA-seq library prep:</p>
<p>. . .</p>
<ul>
<li><p><strong>mRNA enrichment</strong><br>
Needed because mRNA makes up only a few percent of RNA molecules, and done using either:</p>
<ul>
<li><p><strong>PolyA-selection</strong>: specifically selects mature (spliced) mRNAs</p></li>
<li><p><strong>Ribodepletion</strong>: fishes out rRNA — this is suboptimal for mRNA-seq (presence of other RNAs and of pre-mRNAs) but can be a last resort with poor RNA quality</p></li>
</ul></li>
</ul>
<hr style="height:1pt; visibility:hidden;">
<p>. . .</p>
<ul>
<li><p><strong>Library strandedness</strong><br>
Libraries can be “<strong>stranded</strong>” or “<strong>unstranded</strong>” — stranded libraries allow you to tell the directionality of a transcript, which in turn allows you to:</p>
<ul>
<li><p>Distinguish between overlapping genes</p></li>
<li><p>Assess whether you may have DNA contamination</p></li>
</ul></li>
</ul>
</section>
<section id="functional-enrichment-statistics" class="level2" data-number="12">
<h2 data-number="12" class="anchored" data-anchor-id="functional-enrichment-statistics"><span class="header-section-number">12</span> Functional enrichment: statistics</h2>
<p>The two most common statistical approaches are:</p>
<ul>
<li><strong>Overrepresentation Analysis</strong> (<strong>ORA</strong>)<br>
With a list of genes divided in two groups (DE vs.&nbsp;not-DE), test whether the groups contain different frequencies of each functional gene category, like a Chi-Square/Fisher’s Exact Test:</li>
</ul>
<table class="table">
<thead>
<tr class="header">
<th></th>
<th>DE</th>
<th>not-DE</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>In photosynthesis category</td>
<td>15</td>
<td>30</td>
</tr>
<tr class="even">
<td>Not in photosynthesis category</td>
<td>85</td>
<td>14,870</td>
</tr>
</tbody>
</table>
<p>. . .</p>
<hr style="height:1pt; visibility:hidden;">
<ul>
<li><strong>Gene Set Enrichment Analysis</strong> (<strong>GSEA</strong>)<br>
Uses the expression level change (log-fold changes, LFC) of each gene and does not take statistical significance into account. It basically asks whether the LFCs of genes in a category are significantly shifted away from zero, either upregulated or downregulated.</li>
</ul>
</section>
<section id="de-analysis-other-designs" class="level2" data-number="13">
<h2 data-number="13" class="anchored" data-anchor-id="de-analysis-other-designs"><span class="header-section-number">13</span> DE analysis: other designs</h2>
<p>Statistical designs other than pairwise comparisons are also possible. For example:</p>
<ul>
<li><strong>Time series</strong> (or other continuous variables; regression analysis)<br>
Which genes have monotonous in- or decreases across three or more time points?</li>
</ul>
<p>. . .</p>
<hr style="height:1pt; visibility:hidden;">
<ul>
<li><p><strong>Controlling for an independent variable</strong> (fixed or random effect)<br>
Assessing the effect of one variable while controlling for another. E.g.:</p>
<ul>
<li><p>Which genes respond similarly to a virulent (vs.&nbsp;avirulent) pathogen in resistant and susceptible plants?</p></li>
<li><p>Which genes respond similarly to a virulent (vs.&nbsp;avirulent) pathogen regardless of field plot?</p></li>
</ul></li>
</ul>
<p>. . .</p>
<hr style="height:1pt; visibility:hidden;">
<ul>
<li><strong>Interaction effects between two independent variables</strong><br>
E.g.: which genes respond differently to a virulent (vs.&nbsp;avirulent) pathogen in resistant and susceptible plants?</li>
</ul>
</section>
<section id="normalization-by-gene-length" class="level2" data-number="14">
<h2 data-number="14" class="anchored" data-anchor-id="normalization-by-gene-length"><span class="header-section-number">14</span> Normalization by gene length</h2>
<p>Several metrics of gene expression levels exist:</p>
<ul>
<li><p><strong>Adjusted by gene length and library size</strong><br>
E.g., FPKM (superseded) and TPM (Transcripts Per Million)</p></li>
<li><p><strong>Raw counts</strong><br>
These are used by most downstream approaches (more later).</p></li>
</ul>
<hr style="height:1pt; visibility:hidden;">
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
What about gene lengths?
</div>
</div>
<div class="callout-body-container callout-body">
<p>Even though longer genes are more likely to be sampled and gene counts are therefore confounded by gene length, there is no normalization by gene length, because genes are not compared.</p>
</div>
</div>
</section>
<section id="alternative-splicing-isoforms" class="level2" data-number="15">
<h2 data-number="15" class="anchored" data-anchor-id="alternative-splicing-isoforms"><span class="header-section-number">15</span> Alternative splicing &amp; isoforms</h2>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img_rnaseq/DNA_alternative_splicing2.png" class="img-fluid quarto-figure quarto-figure-center figure-img"></p>
<figcaption>https://en.wikipedia.org/wiki/Protein_isoform</figcaption>
</figure>
</div>
</section>
<section id="alternative-splicing-isoforms-1" class="level2" data-number="16">
<h2 data-number="16" class="anchored" data-anchor-id="alternative-splicing-isoforms-1"><span class="header-section-number">16</span> Alternative splicing &amp; isoforms</h2>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Transcript-level vs.&nbsp;gene-level
</div>
</div>
<div class="callout-body-container callout-body">
<p>The terminology “transcript-level” vs.&nbsp;“gene-level”, e.g.&nbsp;in “transcript-level counts” refers to the distinction between having separate counts for each isoform (AKA transcript) versus a single count for each gene. Gene-level counts may either have been aggregated across isoforms, or reads were never assigned to isoforms in the first place.</p>
</div>
</div>
</section>
<section id="what-if-i-dont-have-a-reference-genome" class="level2" data-number="17">
<h2 data-number="17" class="anchored" data-anchor-id="what-if-i-dont-have-a-reference-genome"><span class="header-section-number">17</span> What if I don’t have a reference genome?</h2>
<p>Without a reference genome, an RNA-seq project is considerably more work.<br>
These two extra steps are needed:</p>
<hr style="height:1pt; visibility:hidden;">
<ul>
<li><strong><em>De novo</em></strong> <strong>transcriptome assembly</strong><br>
First, your reads can be used to first assemble a transcriptome from scratch with a tool like <code>Trinity</code>. (This is quite compute-intensive!)</li>
</ul>
<p>. . .</p>
<ul>
<li><p><strong>Transcriptome annotation</strong><br>
Then, your transcriptome will need to be functionally annotated:</p>
<ul>
<li><p>Trying to assign gene names and functional categories to your transcripts</p></li>
<li><p>A very similar process to genome annotation, and works in part by using BLAST or related tools to find orthologous, already annotated, genes in other organisms.</p></li>
</ul></li>
</ul>
<hr style="height:1pt; visibility:hidden;">
<p>. . .</p>
<p>Then, you can <strong>align to your newly assembled transcriptome</strong> the same way you would align to a reference transcriptome (obviously, you won’t have the option to align to the genome).</p>
</section>
<section id="reads-to-counts-alignment-to-a-reference-genome" class="level2" data-number="18">
<h2 data-number="18" class="anchored" data-anchor-id="reads-to-counts-alignment-to-a-reference-genome"><span class="header-section-number">18</span> Reads to counts: alignment to a reference genome</h2>
<p>The <strong>alignment of reads to a reference genome</strong> needs to be <em>“splice-aware”</em>.<br>
Alternatively, you can <em>align to the transcriptome</em> (i.e., all mature transcripts).</p>
<p><br></p>
<p>Alignment usually consists of <strong>two steps</strong>:</p>
<ul>
<li><p>One-time creation of a tool-specific “<strong>index</strong>” of the reference genome/transcriptome.</p></li>
<li><p>For each sample independently, <strong>alignment</strong> of the reads to the reference index.</p></li>
</ul>
<p><br></p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Alignment tools
</div>
</div>
<div class="callout-body-container callout-body">
<p>Some of the currently most commonly used alignment tools are <code>HISAT2</code> &amp; <code>STAR</code> (map to genome), and <code>RSEM</code> &amp; <code>Salmon</code> &amp; <code>Kallisto</code> (map to transcriptome).</p>
</div>
</div>
</section>
<section id="reads-to-counts-alignment-qc" class="level2" data-number="19">
<h2 data-number="19" class="anchored" data-anchor-id="reads-to-counts-alignment-qc"><span class="header-section-number">19</span> Reads to counts: alignment QC</h2>
<p>These kind of stats can only be assessed after aligning to the genome (vs.&nbsp;to the transcriptome).</p>
<ul>
<li><p><strong>Alignment rates</strong><br>
What percentage of reads was successfully aligned?</p>
<ul>
<li><p><strong>Low rates</strong> (&lt;80%) =&gt; cross-species contamination or poor reference genome quality</p></li>
<li><p><strong>Multi-mapped reads</strong> (mapping to multiple locations in the genome) are typically accepted by the mapper but are not or probabilistically quantified — more on this later.</p></li>
</ul></li>
</ul>
<hr style="height:1pt; visibility:hidden;">
<p>. . .</p>
<ul>
<li><p><strong>Alignment targets</strong><br>
What percentages of aligned reads mapped to <em>exons</em> vs.&nbsp;<em>introns</em> vs.&nbsp;<em>intergenic regions</em>?</p>
<ul>
<li><p>High <strong>intronic</strong> mapping rates =&gt; presence of pre-mRNA</p></li>
<li><p>High <strong>intergenic</strong> mapping rates =&gt; DNA contamination or poor genome assembly/annotation quality</p></li>
</ul></li>
</ul>
</section>
<section id="reads-to-counts-alignment-qc-1" class="level2" data-number="20">
<h2 data-number="20" class="anchored" data-anchor-id="reads-to-counts-alignment-qc-1"><span class="header-section-number">20</span> Reads to counts: alignment QC</h2>
<ul>
<li><strong>Across-gene coverage distribution</strong><br>
An uneven distribution may indicate high levels of RNA degradation</li>
</ul>
<hr style="height:1pt; visibility:hidden;">
<ul>
<li><strong>Outlier samples</strong><br>
Are patterns consistent across samples or are there outliers that may need to be removed?</li>
</ul>
<p><br></p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Alignment QC tools
</div>
</div>
<div class="callout-body-container callout-body">
<p>You can get these stats by checking the aligner’s “log” outputs, and by running tools like <code>QualiMap</code> and <code>RSeQC</code>.</p>
</div>
</div>
</section>
<section id="reads-to-counts-read-pre-processing" class="level2" data-number="21">
<h2 data-number="21" class="anchored" data-anchor-id="reads-to-counts-read-pre-processing"><span class="header-section-number">21</span> Reads to counts: Read pre-processing</h2>
<p>Read pre-processing includes the following steps:</p>
<hr style="height:1pt; visibility:hidden;">
<ul>
<li><p><strong>Checking the</strong> <strong>quantity and quality of your reads</strong></p>
<ul>
<li><p>Does not change your data, but helps decide next steps / sample exclusion</p></li>
<li><p>Also useful to check for contamination, library complexity, and adapter content</p></li>
<li><p>Typically done with the tool <code>FastQC</code> (and <code>MultiQC</code> for across-sample summaries)</p></li>
</ul></li>
</ul>
<p>. . .</p>
<hr style="height:1pt; visibility:hidden;">
<ul>
<li><strong>Removing unwanted sequences</strong>
<ul>
<li><strong>Adapters</strong>, <strong>low-quality bases</strong>, and very short reads — with a tool like <code>TrimGalore</code></li>
<li><strong>rRNA</strong>-derived reads — with <code>SortMeRNA</code> (<em>optional</em>)</li>
<li><strong>Contaminant</strong> sequences — with a taxonomic read classification program like <code>Kraken2</code> (<em>optional</em>)</li>
</ul></li>
</ul>
</section>
<section id="functional-enrichment" class="level2" data-number="22">
<h2 data-number="22" class="anchored" data-anchor-id="functional-enrichment"><span class="header-section-number">22</span> Functional enrichment</h2>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Retrieving annotations
</div>
</div>
<div class="callout-body-container callout-body">
<p>Genome annotations for a specific organism often but not always include GO or KEGG terms.</p>
<p>KEGG annotations are more commonly missing but can also be more easily generated, by uploading a FASTA file with amino acid sequences to KEGG’s <code>GhostKOALA</code> webservice: <a href="https://www.kegg.jp/ghostkoala/" class="uri">https://www.kegg.jp/ghostkoala/</a>.</p>
</div>
</div>
</section>
<section id="de-analysis-1-pairwise-comparison" class="level2" data-number="23">
<h2 data-number="23" class="anchored" data-anchor-id="de-analysis-1-pairwise-comparison"><span class="header-section-number">23</span> DE analysis: &gt;1 pairwise comparison</h2>
<p><strong>More than two groups</strong></p>
<div class="columns">
<div class="column" style="width:50%;">
<p>Plants exposed to:</p>
<ul>
<li>A virulent pathogen</li>
<li>An avirulent pathogen</li>
<li>A mock inoculation</li>
</ul>
</div><div class="column" style="width:50%;">
<p>Pairwise comparisons:</p>
<ul>
<li>Virulent vs.&nbsp;avirulent</li>
<li>Mock vs.&nbsp;virulent</li>
<li>Mock vs.&nbsp;avirulent</li>
</ul>
</div>
</div>
<p><br></p>
<p>. . .</p>
<p><strong>More than one independent variable</strong></p>
<div class="columns">
<div class="column" style="width:50%;">
<ul>
<li>Plant type: resistant &amp; susceptible</li>
<li>Pathogen type: virulent &amp; avirulent</li>
</ul>
</div><div class="column" style="width:50%;">
<ul>
<li>Resistant plant vs.&nbsp;virulent pathogen</li>
<li>Resistant plant vs.&nbsp;avirulent pathogen</li>
<li>Susceptible plant vs.&nbsp;virulent pathogen</li>
<li>Susceptible plant vs.&nbsp;avirulent pathogen</li>
</ul>
</div>
</div>
</section>
<section id="side-note-dual-rna-seq" class="level2" data-background-color="mintcream" data-number="24">
<h2 data-background-color="mintcream" data-number="24" class="anchored" data-anchor-id="side-note-dual-rna-seq"><span class="header-section-number">24</span> Side note: Dual RNA-seq</h2>
<p>In agricultural research, it is common to study both a plant host and its adversary.</p>
<p>If the adversary is a microbial/viral/fungal/oomycete pathogen, it is possible or necessary to collect samples that contain both host and pathogen.</p>
<p><br></p>
<p>When these samples are subjected to RNA-seq, they will contain both host and pathogen reads.</p>
<p>These reads can be separated bioinformatically and when both host and pathogen reads are analyzed, this is called a <strong>dual RNA-seq experiment</strong>.</p>


</section>

<a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>