<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.549">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Jelmer Poelstra">
<meta name="dcterms.date" content="2024-02-02">

<title>ENTMLGY 6703 - Jelmer’s lectures - Week 4 lab: RNA-seq differential expression analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">ENTMLGY 6703 - Jelmer’s lectures</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html"> <i class="bi bi-house" role="img">
</i> 
<span class="menu-text"> </span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-week-3-hts" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Week 3: HTS</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-week-3-hts">    
        <li>
    <a class="dropdown-item" href="../lectures/sequencing.html">
 <span class="dropdown-text">Lecture: High-throughput Sequencing &amp; Genomes</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../labs/infrastructure.html">
 <span class="dropdown-text">Lab part 1: Computational Infrastructure</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../labs/data.html">
 <span class="dropdown-text">Lab part 2: Working with genome and HTS Data</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-week-4-rna-seq" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Week 4: RNA-seq</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-week-4-rna-seq">    
        <li class="dropdown-header">Lecture: RNA-seq</li>
        <li class="dropdown-header">Lab: Differential expression analysis</li>
    </ul>
  </li>
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-other-pages" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Other Pages</span>
    </a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-other-pages">    
        <li>
    <a class="dropdown-item" href="../r_homework.html">
 <span class="dropdown-text">R Homework</span></a>
  </li>  
        <li class="dropdown-header">Overview of common Unix shell commands</li>
    </ul>
  </li>
  <li class="nav-item">
    <a class="nav-link" href="../about.html"> 
<span class="menu-text">About</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label=""><i class="bi bi-github"></i></a>
      <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="quarto-navigation-tool-dropdown-0">
          <li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://github.com/jelmerp/ENTMLGY6703">
            Source Code
            </a>
          </li>
          <li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://github.com/jelmerp/ENTMLGY6703/issues/new">
            Report an Issue
            </a>
          </li>
      </ul>
    </div>
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction"><span class="header-section-number">1</span> Introduction</a></li>
  <li><a href="#getting-set-up" id="toc-getting-set-up" class="nav-link" data-scroll-target="#getting-set-up"><span class="header-section-number">2</span> Getting set up</a>
  <ul class="collapse">
  <li><a href="#start-an-rstudio-session-at-osc" id="toc-start-an-rstudio-session-at-osc" class="nav-link" data-scroll-target="#start-an-rstudio-session-at-osc"><span class="header-section-number">2.1</span> Start an RStudio session at OSC</a></li>
  <li><a href="#change-two-settings" id="toc-change-two-settings" class="nav-link" data-scroll-target="#change-two-settings"><span class="header-section-number">2.2</span> Change two settings</a></li>
  <li><a href="#create-a-new-rstudio-project" id="toc-create-a-new-rstudio-project" class="nav-link" data-scroll-target="#create-a-new-rstudio-project"><span class="header-section-number">2.3</span> Create a new RStudio Project</a></li>
  <li><a href="#create-an-r-script" id="toc-create-an-r-script" class="nav-link" data-scroll-target="#create-an-r-script"><span class="header-section-number">2.4</span> Create an R script</a></li>
  <li><a href="#load-the-necessary-packages" id="toc-load-the-necessary-packages" class="nav-link" data-scroll-target="#load-the-necessary-packages"><span class="header-section-number">2.5</span> Load the necessary packages</a></li>
  <li><a href="#define-our-input-files" id="toc-define-our-input-files" class="nav-link" data-scroll-target="#define-our-input-files"><span class="header-section-number">2.6</span> Define our input files</a></li>
  </ul></li>
  <li><a href="#create-a-deseq2-object" id="toc-create-a-deseq2-object" class="nav-link" data-scroll-target="#create-a-deseq2-object"><span class="header-section-number">3</span> Create a DESeq2 object</a>
  <ul class="collapse">
  <li><a href="#metadata" id="toc-metadata" class="nav-link" data-scroll-target="#metadata"><span class="header-section-number">3.1</span> Metadata</a></li>
  <li><a href="#count-table" id="toc-count-table" class="nav-link" data-scroll-target="#count-table"><span class="header-section-number">3.2</span> Count table</a></li>
  <li><a href="#create-the-deseq2-object" id="toc-create-the-deseq2-object" class="nav-link" data-scroll-target="#create-the-deseq2-object"><span class="header-section-number">3.3</span> Create the DESeq2 object</a></li>
  </ul></li>
  <li><a href="#exploratory-data-analysis" id="toc-exploratory-data-analysis" class="nav-link" data-scroll-target="#exploratory-data-analysis"><span class="header-section-number">4</span> Exploratory Data Analysis</a>
  <ul class="collapse">
  <li><a href="#our-count-matrix" id="toc-our-count-matrix" class="nav-link" data-scroll-target="#our-count-matrix"><span class="header-section-number">4.1</span> Our count matrix</a></li>
  <li><a href="#principal-component-analysis-pca" id="toc-principal-component-analysis-pca" class="nav-link" data-scroll-target="#principal-component-analysis-pca"><span class="header-section-number">4.2</span> Principal Component Analysis (PCA)</a></li>
  </ul></li>
  <li><a href="#differential-expression-de-analysis" id="toc-differential-expression-de-analysis" class="nav-link" data-scroll-target="#differential-expression-de-analysis"><span class="header-section-number">5</span> Differential Expression (DE) analysis</a>
  <ul class="collapse">
  <li><a href="#figuring-out-how-to-do-the-analysis" id="toc-figuring-out-how-to-do-the-analysis" class="nav-link" data-scroll-target="#figuring-out-how-to-do-the-analysis"><span class="header-section-number">5.1</span> Figuring out how to do the analysis</a></li>
  <li><a href="#setting-the-statistical-design" id="toc-setting-the-statistical-design" class="nav-link" data-scroll-target="#setting-the-statistical-design"><span class="header-section-number">5.2</span> Setting the statistical design</a></li>
  <li><a href="#running-the-de-analysis" id="toc-running-the-de-analysis" class="nav-link" data-scroll-target="#running-the-de-analysis"><span class="header-section-number">5.3</span> Running the DE analysis</a></li>
  </ul></li>
  <li><a href="#extracting-the-de-results" id="toc-extracting-the-de-results" class="nav-link" data-scroll-target="#extracting-the-de-results"><span class="header-section-number">6</span> Extracting the DE results</a>
  <ul class="collapse">
  <li><a href="#the-results-table" id="toc-the-results-table" class="nav-link" data-scroll-target="#the-results-table"><span class="header-section-number">6.1</span> The results table</a></li>
  </ul></li>
  <li><a href="#visualizing-the-de-results" id="toc-visualizing-the-de-results" class="nav-link" data-scroll-target="#visualizing-the-de-results"><span class="header-section-number">7</span> Visualizing the DE results</a>
  <ul class="collapse">
  <li><a href="#volcano-plot" id="toc-volcano-plot" class="nav-link" data-scroll-target="#volcano-plot"><span class="header-section-number">7.1</span> Volcano plot</a></li>
  <li><a href="#plot-specific-genes" id="toc-plot-specific-genes" class="nav-link" data-scroll-target="#plot-specific-genes"><span class="header-section-number">7.2</span> Plot specific genes</a></li>
  <li><a href="#heatmaps" id="toc-heatmaps" class="nav-link" data-scroll-target="#heatmaps"><span class="header-section-number">7.3</span> Heatmaps</a></li>
  </ul></li>
  <li><a href="#in-closing" id="toc-in-closing" class="nav-link" data-scroll-target="#in-closing"><span class="header-section-number">8</span> In Closing</a>
  <ul class="collapse">
  <li><a href="#next-steps" id="toc-next-steps" class="nav-link" data-scroll-target="#next-steps"><span class="header-section-number">8.1</span> Next steps</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Week 4 lab: RNA-seq differential expression analysis</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Jelmer Poelstra </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">February 2, 2024</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<p><br></p>
<section id="introduction" class="level2" data-number="1">
<h2 data-number="1" class="anchored" data-anchor-id="introduction"><span class="header-section-number">1</span> Introduction</h2>
<p>Last week, you explored the RNA-seq reads from the 2023 Molecular Ecology paper <a href="https://doi.org/10.1111/mec.17240">“<em>Two avian Plasmodium species trigger different transcriptional responses on their vector Culex pipiens</em>”</a>, as well as the <em>Culex pipiens</em> reference genome files.</p>
<p>In yesterday’s lecture, you learned about the steps to generate a gene count table from this input data (FASTQ reads + reference genome files), but as mentioned, we will not go through those steps ourselves.</p>
<p>Instead, I have generated a gene count table for you, and we will now be doing differential expression analysis in R, using the popular <strong>DESeq2</strong> package (<a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8">paper</a>, <a href="https://bioconductor.org/packages/release/bioc/html/DESeq2.html">website</a>).</p>
<p><br></p>
</section>
<section id="getting-set-up" class="level2" data-number="2">
<h2 data-number="2" class="anchored" data-anchor-id="getting-set-up"><span class="header-section-number">2</span> Getting set up</h2>
<section id="start-an-rstudio-session-at-osc" class="level3" data-number="2.1">
<h3 data-number="2.1" class="anchored" data-anchor-id="start-an-rstudio-session-at-osc"><span class="header-section-number">2.1</span> Start an RStudio session at OSC</h3>
<ol type="1">
<li><p>Log in to OSC at <a href="https://ondemand.osc.edu" class="uri">https://ondemand.osc.edu</a></p></li>
<li><p>Click on <strong><code>Interactive Apps</code></strong> (top bar) and then <strong><code>RStudio Server</code></strong> (all the way at the bottom)</p></li>
<li><p>Fill out the form as follows:</p>
<ul>
<li>Cluster: <strong><code>Pitzer</code></strong></li>
<li>R version: <strong><code>4.3.0</code></strong></li>
<li>Project: <strong><code>PAS2250</code></strong></li>
<li>Number of hours: <strong><code>4</code></strong></li>
<li>Node type: <strong><code>any</code></strong></li>
<li>Number of cores: <strong><code>2</code></strong></li>
</ul></li>
</ol>
<details>
<summary>
<em>Click to see a screenshot</em>
</summary>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img_de/rstudio_form.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:50.0%"></p>
</figure>
</div>
</details>
<ol start="4" type="1">
<li><p>Click the big blue <strong><code>Launch</code></strong> button at the bottom</p></li>
<li><p>Now, you should be sent to a new page with a box at the top for your RStudio Server “job”, which should initially be “Queued” (waiting to start).</p></li>
</ol>
<details>
<summary>
<em>Click to see a screenshot</em>
</summary>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img_de/rstudio_queued.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:70.0%"></p>
</figure>
</div>
</details>
<ol start="6" type="1">
<li>Your job should start running very soon, with the top bar of the box turning green and saying “Running”.</li>
</ol>
<details>
<summary>
<em>Click to see a screenshot</em>
</summary>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img_de/rstudio_running.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:70.0%"></p>
</figure>
</div>
</details>
<ol start="7" type="1">
<li>Click <strong><code>Connect to RStudio Server</code></strong> at the bottom of the box, and an RStudio Server instance will open in a new browser tab. You’re ready to go!</li>
</ol>
<hr style="height:1pt; visibility:hidden;">
</section>
<section id="change-two-settings" class="level3" data-number="2.2">
<h3 data-number="2.2" class="anchored" data-anchor-id="change-two-settings"><span class="header-section-number">2.2</span> Change two settings</h3>
<p><em>First</em>, we’ll prevent R from saving your <strong>“Workspace”</strong>:</p>
<ol type="1">
<li>Click <strong><code>Tools</code></strong> (top bar, below your browser’s address bar) &gt; <strong><code>Global Options</code></strong></li>
<li>In the pop-up window (stay on the <code>General</code> tab), change the settings under the “Workspace” heading to:</li>
</ol>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img_de/workspace.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:45.0%"></p>
</figure>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Why are we doing this? <em>(Click to expand)</em>
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>In short, the default behavior of saving and restoring your “Workspace”, which are all the items (objects) that you create during an R session, is bad practice. Instead, you should recreate your environment from a script and/or saved files with individual pieces of data, as we’ll do today.</p>
</div>
</div>
</div>
<hr style="height:1pt; visibility:hidden;">
<p><em>Second</em>, we’ll “update” our <strong>pipe symbol</strong> from <code>%&gt;%</code> (the magrittr pipe that requires a package to work) to <code>|&gt;</code> (the new base R pipe that does not require a package):</p>
<ol type="1">
<li>Again click <strong><code>Tools</code></strong> &gt; <strong><code>Global Options</code></strong> (you may still be there)</li>
<li>Now go to <strong><code>Code</code></strong> tab in the side panel on the left, and check the box for <code>Use native pipe operator, |&gt; (requires R 4.1+)</code></li>
<li>Click <strong><code>OK</code></strong> at the bottom of the pop-up window</li>
</ol>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img_de/rstudio_pipe.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:70.0%"></p>
</figure>
</div>
<hr style="height:1pt; visibility:hidden;">
</section>
<section id="create-a-new-rstudio-project" class="level3" data-number="2.3">
<h3 data-number="2.3" class="anchored" data-anchor-id="create-a-new-rstudio-project"><span class="header-section-number">2.3</span> Create a new RStudio Project</h3>
<p>Using an “RStudio Project” will most of all help to make sure your working directory in R is correct.</p>
<p>To create a new RStudio Project inside your personal dir in <code>/fs/scratch/PAS2250/ENT6703</code>:</p>
<ol type="1">
<li>Click <strong><code>File</code></strong> (top bar, below your browser’s address bar) &gt; <strong><code>New Project</code></strong></li>
<li>In the popup window, click <strong><code>Existing Directory</code></strong>.</li>
</ol>
<details>
<summary>
<em>Click to see a screenshot</em>
</summary>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img_de/rstudio_proj_existingdir.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:40.0%"></p>
</figure>
</div>
</details>
<ol start="3" type="1">
<li>Click <strong><code>Browse...</code></strong> to select your personal dir.</li>
</ol>
<details>
<summary>
<em>Click to see a screenshot</em>
</summary>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img_de/rstudio_proj_browse.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:40.0%"></p>
</figure>
</div>
</details>
<ol start="4" type="1">
<li>In the next window, you should be in your Home directory (abbreviated as <strong><code>~</code></strong>), from which you can’t click your way to <code>/fs/scratch</code>! Instead, you’ll first have to click on the (very small!) <strong><code>...</code></strong> highlighted in the screenshot below:</li>
</ol>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img_de/rstudio_proj_dotdotdot_ed.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:50.0%"></p>
</figure>
</div>
<ol start="5" type="1">
<li>Type at least part of the path to your personal dir (which is in <code>/fs/scratch/PAS2250/ENT6703</code>), e.g.&nbsp;as shown below, and click <strong><code>OK</code></strong>:</li>
</ol>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img_de/rstudio_proj_path.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:35.0%"></p>
</figure>
</div>
<ol start="6" type="1">
<li>Now you should be able to browse/click the rest of the way to your personal directory.</li>
<li>Click <strong><code>Choose</code></strong> to pick your selected directory.</li>
<li>Click <strong><code>Create Project</code></strong>.</li>
</ol>
<hr style="height:1pt; visibility:hidden;">
</section>
<section id="create-an-r-script" class="level3" data-number="2.4">
<h3 data-number="2.4" class="anchored" data-anchor-id="create-an-r-script"><span class="header-section-number">2.4</span> Create an R script</h3>
<p>We’re going to write all our code in an R script instead of typing it directly in the console. This helps us to <strong>keep track of what we’ve been doing</strong>, especially in the longer run, and to be able to <strong>re-run our code</strong> after modifying input data or one of the lines of code.</p>
<p>Create and open a new R script by clicking <strong><code>File</code></strong> (top menu bar) &gt; <strong><code>New File</code></strong> &gt; <strong><code>R Script</code></strong>.</p>
<p>Save this new script right away by clicking <strong><code>File</code></strong> &gt; <strong><code>Save As</code></strong>, then click <strong><code>New Folder</code></strong> and create a folder named “scripts”. Inside that folder, save the script with a name like <strong><code>lab4_DE.R</code></strong> (the extension for R scripts is <code>.R</code>).</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Make sure to type all the R code below inside your script, and then send it to the console from there.
</div>
</div>
<div class="callout-body-container callout-body">
<p>You can send code to the console by pressing <kbd><strong>Ctrl</strong></kbd> <strong>+</strong> <kbd><strong>Enter</strong></kbd> on Windows, or <kbd>Cmd</kbd> + <kbd>Return</kbd> on a Mac.</p>
</div>
</div>
<hr style="height:1pt; visibility:hidden;">
</section>
<section id="load-the-necessary-packages" class="level3" data-number="2.5">
<h3 data-number="2.5" class="anchored" data-anchor-id="load-the-necessary-packages"><span class="header-section-number">2.5</span> Load the necessary packages</h3>
<p>In R, we need to use “packages” (basically, add-ons) to perform specialized tasks like differential expression analysis<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a>. This is quite straightforward in principle, but at OSC, there can be some hickups and at the very least, it will take much longer to install packages than on your own computer<a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a>.</p>
<p>I have therefore created a “library” (a directory with a collection of packages) for you — you can load the packages from that library, without needing to install them yourself. Type the code below in your R script and send it to the R console:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>custom_library <span class="ot">&lt;-</span> <span class="st">"/fs/scratch/PAS2250/ENT6703/share/rlib"</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(withr, <span class="at">lib.loc =</span> custom_library)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr, <span class="at">lib.loc =</span> custom_library)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse, <span class="at">lib.loc =</span> custom_library)       <span class="co"># Misc. data manipulation and plotting</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DESeq2, <span class="at">lib.loc =</span> custom_library)          <span class="co"># Differential expression analysis</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(EnhancedVolcano, <span class="at">lib.loc =</span> custom_library) <span class="co"># Volcano plot</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(pheatmap, <span class="at">lib.loc =</span> custom_library)        <span class="co"># Heatmap plot</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr style="height:1pt; visibility:hidden;">
</section>
<section id="define-our-input-files" class="level3" data-number="2.6">
<h3 data-number="2.6" class="anchored" data-anchor-id="define-our-input-files"><span class="header-section-number">2.6</span> Define our input files</h3>
<p>For the differential expression analysis, we have the following input files:</p>
<ul>
<li><strong>Metadata table</strong> — The metadata we saw last week, to enable between-treatment comparisons</li>
<li><strong>Gene counts table</strong> — Produced by running the <a href="https://nf-co.re/rnaseq">nf-core rnaseq workflow</a> on the input data we saw last week</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>count_table_file <span class="ot">&lt;-</span> <span class="st">"../share/results/counts/salmon.merged.gene_counts_length_scaled.tsv"</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>metadata_file <span class="ot">&lt;-</span> <span class="st">"../share/data/meta/metadata.tsv"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
</section>
</section>
<section id="create-a-deseq2-object" class="level2" data-number="3">
<h2 data-number="3" class="anchored" data-anchor-id="create-a-deseq2-object"><span class="header-section-number">3</span> Create a DESeq2 object</h2>
<p>The DESeq2 package has its own “object type” (format) and before we can do anything else, we need to create a DESeq2 object from three components:</p>
<ol type="1">
<li><strong>Metadata</strong><br>
We should have all of our independent variables in columns in the metadata, which will allow DESeq2 to compare groups of samples</li>
<li><strong>Count table</strong><br>
DESeq2 will of course need gene counts for its analysis. This table has one row per gene, and one column per sample. The sample IDs in the column names should match with those in the metadata.</li>
<li><strong>A statistical design</strong><br>
Instead of telling DESeq2 about the statistical design (i.e., which groups to compare) on the fly when we run the analysis, we need to include this information in the object itself.</li>
</ol>
<hr style="height:1pt; visibility:hidden;">
<section id="metadata" class="level3" data-number="3.1">
<h3 data-number="3.1" class="anchored" data-anchor-id="metadata"><span class="header-section-number">3.1</span> Metadata</h3>
<p>First, we’ll load the metadata file and take a look at the resulting dataframe:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read in the count table</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>meta_raw <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(metadata_file, <span class="at">show_col_types =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Take a look at the first 6 rows</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(meta_raw)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<details>
<summary>
<em>Click to see the result</em>
</summary>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 3
  sample_id   time  treatment  
  &lt;chr&gt;       &lt;chr&gt; &lt;chr&gt;      
1 ERR10802882 10dpi cathemerium
2 ERR10802875 10dpi cathemerium
3 ERR10802879 10dpi cathemerium
4 ERR10802883 10dpi cathemerium
5 ERR10802878 10dpi control    
6 ERR10802884 10dpi control    </code></pre>
</div>
</div>
</details>
<p>We’ll make sure the data frame is sorted by sample ID, and that the sample IDs are contained in “row names”:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>meta <span class="ot">&lt;-</span> meta_raw <span class="sc">|&gt;</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Sort by the 'sample_id' column</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(sample_id) <span class="sc">|&gt;</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Turn the 'sample_id' column into row names:</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">column_to_rownames</span>(<span class="st">"sample_id"</span>) <span class="sc">|&gt;</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Turn the 'time' and 'treatment' columns into "factors":</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">time =</span> <span class="fu">factor</span>(time, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"24hpi"</span>, <span class="st">"10dpi"</span>)),</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>         <span class="at">treatment =</span> <span class="fu">factor</span>(treatment, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"control"</span>, <span class="st">"cathemerium"</span>, <span class="st">"relictum"</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(meta)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<details>
<summary>
<em>Click to see the result</em>
</summary>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>             time   treatment
ERR10802863 24hpi     control
ERR10802864 24hpi cathemerium
ERR10802865 24hpi    relictum
ERR10802866 24hpi     control
ERR10802867 24hpi cathemerium
ERR10802868 24hpi    relictum</code></pre>
</div>
</div>
</details>
<p>We changed the two columns with the independent variables (<code>time</code> and <code>treatment</code>) into factors, because DESEq2 wants this — and this also allowed us to use a custom, non-alphanumeric ordering where <code>24hpi</code> comes before <code>10dpi</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(meta<span class="sc">$</span>time)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 24hpi 24hpi 24hpi 24hpi 24hpi 24hpi
Levels: 24hpi 10dpi</code></pre>
</div>
</div>
<hr style="height:1pt; visibility:hidden;">
</section>
<section id="count-table" class="level3" data-number="3.2">
<h3 data-number="3.2" class="anchored" data-anchor-id="count-table"><span class="header-section-number">3.2</span> Count table</h3>
<p>Second, we will load the count table into R:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read in the count table</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>count_df <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(count_table_file, <span class="at">show_col_types =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Take a look at the first 6 rows</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(count_df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<details>
<summary>
<em>Click to see the result</em>
</summary>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 24
  gene_id gene_name ERR10802863 ERR10802864 ERR10802865 ERR10802866 ERR10802867
  &lt;chr&gt;   &lt;chr&gt;           &lt;dbl&gt;       &lt;dbl&gt;       &lt;dbl&gt;       &lt;dbl&gt;       &lt;dbl&gt;
1 ATP6    ATP6         10275.       8255.       4103.      18615.      11625.  
2 ATP8    ATP8             3.85        2.92        2.33        7.76        7.01
3 COX1    COX1         88041.      83394.      36975.     136054.     130863.  
4 COX2    COX2          8749.       7925.       2901.      16802.      10026.  
5 COX3    COX3         55772.      50312.      35074.      80510.      69850.  
6 CYTB    CYTB         38543.      36352.      22185.      62147.      57461.  
# ℹ 17 more variables: ERR10802868 &lt;dbl&gt;, ERR10802869 &lt;dbl&gt;, ERR10802870 &lt;dbl&gt;,
#   ERR10802871 &lt;dbl&gt;, ERR10802874 &lt;dbl&gt;, ERR10802875 &lt;dbl&gt;, ERR10802876 &lt;dbl&gt;,
#   ERR10802877 &lt;dbl&gt;, ERR10802878 &lt;dbl&gt;, ERR10802879 &lt;dbl&gt;, ERR10802880 &lt;dbl&gt;,
#   ERR10802881 &lt;dbl&gt;, ERR10802882 &lt;dbl&gt;, ERR10802883 &lt;dbl&gt;, ERR10802884 &lt;dbl&gt;,
#   ERR10802885 &lt;dbl&gt;, ERR10802886 &lt;dbl&gt;</code></pre>
</div>
</div>
</details>
<p>We next have to make several modifications, because DESeq2 expects an all-numeric matrix with whole numbers (integers) and gene IDs are “row names” rather than as a separate column:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Prepare the count table so it can be loaded into DESeq2</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>count_mat <span class="ot">&lt;-</span> count_df <span class="sc">|&gt;</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Turn the 'gene_id' column into row names:</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">column_to_rownames</span>(<span class="st">"gene_id"</span>) <span class="sc">|&gt;</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Remove a remaining non-numeric column, with gene names:</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>gene_name) <span class="sc">|&gt;</span></span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># We should round everything to whole numbers:</span></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">round</span>() <span class="sc">|&gt;</span></span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>  <span class="co"># We should convert it to a formal matrix format:</span></span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.matrix</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Take a look at the first 6 rows</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(count_mat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<details>
<summary>
<em>Click to see the result</em>
</summary>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>     ERR10802863 ERR10802864 ERR10802865 ERR10802866 ERR10802867 ERR10802868
ATP6       10275        8255        4103       18615       11625        7967
ATP8           4           3           2           8           7           2
COX1       88041       83394       36975      136054      130863       62279
COX2        8749        7925        2901       16802       10026        6701
COX3       55772       50312       35074       80510       69850       42478
CYTB       38543       36352       22185       62147       57461       28159
     ERR10802869 ERR10802870 ERR10802871 ERR10802874 ERR10802875 ERR10802876
ATP6       12788        4408       13648       13834        1346       10032
ATP8           2           0           2           1           3           2
COX1      109596      106402      104394       77682       38276       78290
COX2       11494        6603       11151        9893        1473       13146
COX3       68228       71945       66900       52368       14665       37275
CYTB       46219       52035       46090       35247       17449       38762
     ERR10802877 ERR10802878 ERR10802879 ERR10802880 ERR10802881 ERR10802882
ATP6         987        1834        3337        5036        1983       11586
ATP8           0           0           0           3           0          27
COX1       17785       32099       64490       63960       50965       76113
COX2        1141        1907        3439        8334        2063       12752
COX3        8797       15948       26278       29997       17802       35419
CYTB       11177       22262       34368       33401       25854       43912
     ERR10802883 ERR10802884 ERR10802885 ERR10802886
ATP6       18821        2792       11749        6682
ATP8          40           0           8           1
COX1      108343       65829      107741       94682
COX2       19148        2713       17947       10656
COX3       51441       24915       50029       47750
CYTB       57844       34616       50587       51198</code></pre>
</div>
</div>
</details>
<section id="check-that-the-sample-ids-match" class="level4">
<h4 class="anchored" data-anchor-id="check-that-the-sample-ids-match">Check that the sample IDs match</h4>
<p>When creating the DESeq2 object, DESeq2 assumes that sample IDs in both tables match and are provided in the same order. Let’s make sure this is indeed the case:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">all</span>(<span class="fu">row.names</span>(meta) <span class="sc">==</span> <span class="fu">colnames</span>(count_mat))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
</div>
<hr style="height:1pt; visibility:hidden;">
</section>
</section>
<section id="create-the-deseq2-object" class="level3" data-number="3.3">
<h3 data-number="3.3" class="anchored" data-anchor-id="create-the-deseq2-object"><span class="header-section-number">3.3</span> Create the DESeq2 object</h3>
<p>We will create the DESeq2 object using the function <code>DESeqDataSetFromMatrix()</code>, which we will provide with three arguments corresponding to the components discussed above:</p>
<ul>
<li>The metadata with argument <strong><code>colData</code></strong>.</li>
<li>The count data with argument <strong><code>countData</code></strong>.</li>
<li>The statistical design for the DE analysis with argument <strong><code>design</code></strong>. For now, we will specify <strong><code>~1</code></strong>, which effectively means “no design” — we will change this before the actual DE analysis.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span>(<span class="at">colData =</span> meta, <span class="at">countData =</span> count_mat, <span class="at">design =</span> <span class="sc">~</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>converting counts to integer mode</code></pre>
</div>
</div>
<hr style="height:1pt; visibility:hidden;">
<p>Before we will run the differential expression analysis, though, we will do a bit of exploratory data analysis.</p>
<p><br></p>
</section>
</section>
<section id="exploratory-data-analysis" class="level2" data-number="4">
<h2 data-number="4" class="anchored" data-anchor-id="exploratory-data-analysis"><span class="header-section-number">4</span> Exploratory Data Analysis</h2>
<section id="our-count-matrix" class="level3" data-number="4.1">
<h3 data-number="4.1" class="anchored" data-anchor-id="our-count-matrix"><span class="header-section-number">4.1</span> Our count matrix</h3>
<p>What are number of rows and columns of the count matrix?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(count_mat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 18855    22</code></pre>
</div>
</div>
<p>How many genes have non-zero counts?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(count_mat[<span class="fu">rowSums</span>(count_mat) <span class="sc">&gt;</span> <span class="dv">0</span>, ])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 17788    22</code></pre>
</div>
</div>
<details>
<summary>
<strong>Your Turn</strong>: How many genes have total counts of at least 10? <em>(Click to see the solution)</em>
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(count_mat[<span class="fu">rowSums</span>(count_mat) <span class="sc">&gt;=</span> <span class="dv">10</span>, ])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 16682    22</code></pre>
</div>
</div>
</details>
<p>How are counts distributed across samples? That is, we would like a sum of counts for each column:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colSums</span>(count_mat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>ERR10802863 ERR10802864 ERR10802865 ERR10802866 ERR10802867 ERR10802868 
   24297245    17177436    22745445    26849403    21471477    17506262 
ERR10802869 ERR10802870 ERR10802871 ERR10802874 ERR10802875 ERR10802876 
   24299398    25490128    26534405    22194841    18927885    28804150 
ERR10802877 ERR10802878 ERR10802879 ERR10802880 ERR10802881 ERR10802882 
    9498249    14807513    20667093    23107463    17545375    19088206 
ERR10802883 ERR10802884 ERR10802885 ERR10802886 
   21418234    19420046    24367372    25452228 </code></pre>
</div>
</div>
<details>
<summary>
<strong>Your Turn</strong>: That’s not that easy to read / interpret. Can you instead get these numbers in millions, rounded to whole numbers, and sorted? <em>(Click to see the solution)</em>
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sort</span>(<span class="fu">round</span>(<span class="fu">colSums</span>(count_mat) <span class="sc">/</span> <span class="dv">1000000</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>ERR10802877 ERR10802878 ERR10802864 ERR10802868 ERR10802881 ERR10802875 
          9          15          17          18          18          19 
ERR10802882 ERR10802884 ERR10802867 ERR10802879 ERR10802883 ERR10802874 
         19          19          21          21          21          22 
ERR10802865 ERR10802880 ERR10802863 ERR10802869 ERR10802885 ERR10802870 
         23          23          24          24          24          25 
ERR10802886 ERR10802866 ERR10802871 ERR10802876 
         25          27          27          29 </code></pre>
</div>
</div>
</details>
<p>We can also simply print the <code>dds</code> object, which will give a summary of its contents:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>dds</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: DESeqDataSet 
dim: 18855 22 
metadata(1): version
assays(1): counts
rownames(18855): ATP6 ATP8 ... unassigned_gene_8 unassigned_gene_9
rowData names(0):
colnames(22): ERR10802863 ERR10802864 ... ERR10802885 ERR10802886
colData names(2): time treatment</code></pre>
</div>
</div>
<details>
<summary>
<strong>Bonus section</strong>: Histograms of gene counts <em>(Click to expand)</em>
</summary>
<p>To plot a histogram of mean gene counts across samples, we’ll first create a data frame with these mean gene counts:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>mean_counts <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(count_mat) <span class="sc">/</span> <span class="fu">ncol</span>(count_mat)</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>mean_counts_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(mean_counts) <span class="sc">|&gt;</span> </span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="st">"gene_id"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we can make the histogram:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(mean_counts_df, <span class="fu">aes</span>(<span class="at">x =</span> mean_counts)) <span class="sc">+</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>() <span class="sc">+</span></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_log10</span>() <span class="sc">+</span></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Transformation introduced infinite values in continuous y-axis</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 19 rows containing missing values (`geom_bar()`).</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="DE_files/figure-html/unnamed-chunk-27-1.png" class="img-fluid figure-img" style="width:85.0%"></p>
</figure>
</div>
</div>
</div>
<p>Zoom in a bit:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(mean_counts_df, <span class="fu">aes</span>(<span class="at">x =</span> mean_counts)) <span class="sc">+</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>() <span class="sc">+</span></span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_log10</span>() <span class="sc">+</span></span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">100000</span>)) <span class="sc">+</span></span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 7 rows containing non-finite values (`stat_bin()`).</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Transformation introduced infinite values in continuous y-axis</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 6 rows containing missing values (`geom_bar()`).</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="DE_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid figure-img" style="width:85.0%"></p>
</figure>
</div>
</div>
</div>
<p>Which genes have the highest counts?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sort</span>(<span class="fu">rowSums</span>(count_mat) <span class="sc">/</span> <span class="fu">ncol</span>(count_mat), <span class="at">decreasing =</span> <span class="cn">TRUE</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      LOC120426760 unassigned_gene_22       LOC120413044       LOC120419254 
         645940.86          552075.86          233832.14          171938.27 
      LOC120413042       LOC120425410       LOC120418794       LOC120415720 
         154915.68          105133.50          101350.32           99153.86 
      LOC120422638       LOC120413219 
          98092.05           89276.14 </code></pre>
</div>
</div>
</details>
<p><br></p>
</section>
<section id="principal-component-analysis-pca" class="level3" data-number="4.2">
<h3 data-number="4.2" class="anchored" data-anchor-id="principal-component-analysis-pca"><span class="header-section-number">4.2</span> Principal Component Analysis (PCA)</h3>
<p>We will not run a Principal Component Analysis (PCA) to look for overall patterns of (dis)similarity among our samples. This will help us answer questions like:</p>
<ul>
<li>Do the samples cluster by treatment (infection status) and/or timepoint?</li>
<li>Which of these two variables has a greater effect on overall levels of gene expression?</li>
<li>Is there an overall interaction between these two variables?</li>
</ul>
<p>First, we will have to normalize the count data to have even sampling across samples (with respect to “library size”, total number of reads per sample) and approximately even variance:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>dds_vst <span class="ot">&lt;-</span> <span class="fu">varianceStabilizingTransformation</span>(dds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
The authors of the study did this as well:
</div>
</div>
<div class="callout-body-container callout-body">
<blockquote class="blockquote">
<p><em>We carried out a Variance Stabilizing Transformation (VST) of the counts to represent the samples on a PCA plot.</em></p>
</blockquote>
</div>
</div>
<p>Next, we run and plot the PCA with a single function call, <code>plotPCA</code> from DESeq2 (the <code>ntop</code> argument refers to the top number of most variable genes to use for the PCA):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotPCA</span>(dds_vst, <span class="at">ntop =</span> <span class="dv">500</span>, <span class="at">intgroup =</span> <span class="fu">c</span>(<span class="st">"time"</span>, <span class="st">"treatment"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<details>
<summary>
<em>Click to see the plot</em>
</summary>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="DE_files/figure-html/unnamed-chunk-32-1.png" class="img-fluid figure-img" style="width:85.0%"></p>
</figure>
</div>
</div>
</div>
</details>
<p>How does this compare to the PCA plot in the paper (their Fig. 1)?</p>
<details>
<summary>
<strong>Your Turn</strong>: Try to answer the three questions asked at the beginning of this PCA section. <em>(Click to see the solution)</em>
</summary>
</details>
<details>
<summary>
<strong>Bonus Exercise</strong>: Compare the plot with different values for <code>ntop</code>
</summary>
</details>
<details>
<summary>
<strong>Bonus Section</strong>: Customizing the PCA plot <em>(Click to expand)</em>
</summary>
<p>We’ll do some customization to make the plot look better. The biggest issue (I think) with the above plot is that each combination of time point and treatment has a distinct color — it would be better to use point colors only to distinguish one of the variables, and point shape to distinguish the others (this is also how it was done in the paper’s Fig. 1).</p>
<p>First, if we use</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>pca_df <span class="ot">&lt;-</span> <span class="fu">plotPCA</span>(dds_vst, <span class="at">ntop =</span> <span class="dv">500</span>, <span class="at">intgroup =</span> <span class="fu">c</span>(<span class="st">"time"</span>, <span class="st">"treatment"</span>), <span class="at">returnData =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We’ll extract the percentage of variance explained by different principal components, so we can later add this information to the plot:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>pct_var <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="dv">100</span> <span class="sc">*</span> <span class="fu">attr</span>(pca_df, <span class="st">"percentVar"</span>), <span class="dv">1</span>)</span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>pct_var</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 85.3  3.1</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(pca_df,</span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> PC1, <span class="at">y =</span> PC2, <span class="at">color =</span> treatment, <span class="at">shape =</span> time)) <span class="sc">+</span></span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">5</span>) <span class="sc">+</span></span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="fu">paste0</span>(<span class="st">"PC1 ("</span>, pct_var[<span class="dv">1</span>], <span class="st">"%)"</span>),</span>
<span id="cb50-5"><a href="#cb50-5" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="fu">paste0</span>(<span class="st">"PC2 ("</span>, pct_var[<span class="dv">2</span>], <span class="st">"%)"</span>)) <span class="sc">+</span></span>
<span id="cb50-6"><a href="#cb50-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_brewer</span>(<span class="at">palette =</span> <span class="st">"Dark2"</span>, <span class="at">name =</span> <span class="st">"Infection status"</span>) <span class="sc">+</span></span>
<span id="cb50-7"><a href="#cb50-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_shape</span>(<span class="at">name =</span> <span class="st">"Time points"</span>) <span class="sc">+</span></span>
<span id="cb50-8"><a href="#cb50-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_fixed</span>() <span class="sc">+</span></span>
<span id="cb50-9"><a href="#cb50-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb50-10"><a href="#cb50-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.grid.minor =</span> <span class="fu">element_blank</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="DE_files/figure-html/unnamed-chunk-35-1.png" class="img-fluid figure-img" style="width:85.0%"></p>
</figure>
</div>
</div>
</div>
</details>
<p><br></p>
</section>
</section>
<section id="differential-expression-de-analysis" class="level2" data-number="5">
<h2 data-number="5" class="anchored" data-anchor-id="differential-expression-de-analysis"><span class="header-section-number">5</span> Differential Expression (DE) analysis</h2>
<section id="figuring-out-how-to-do-the-analysis" class="level3" data-number="5.1">
<h3 data-number="5.1" class="anchored" data-anchor-id="figuring-out-how-to-do-the-analysis"><span class="header-section-number">5.1</span> Figuring out how to do the analysis</h3>
<p>First, let’s see how the DE analysis was done in the paper:</p>
<blockquote class="blockquote">
<p><em>Then, we used the DESeq2 package (Love et al., 2014) to perform the differential gene expression analysis comparing: (i) P. relictum-infected mosquitoes vs.&nbsp;controls, (ii) P. cathemerium-infected mosquitoes vs.&nbsp;controls, and (iii) P. relictum-infected mosquitoes vs.&nbsp;P. cathemerium-infected mosquitoes.</em></p>
</blockquote>
<p>This is not terribly detailed and could be interpreted in a couple of different ways. For example, they may have compared infection statuses by ignoring time points or by controlling for time points (and their are different ways to do the latter).</p>
<p>Ignoring time would mean analyzing the full dataset (all time points) while only using the infection status as an independent variable, i.e.&nbsp;the design <code>~treatment</code>.</p>
<details>
<summary>
<strong>Your Turn</strong>: Given the PCA results, do you think that ignoring time is a good idea? <em>(Click to see the solution)</em>
</summary>
<p>Nope!</p>
</details>
<p>Controlling for time can additionally be done in two ways:</p>
<ul>
<li>A two-factor analysis: <code>~ time + treatment</code></li>
<li>Pairwise comparisons between each combination of time and treatment (we’ll see below <em>how</em> we can do that)</li>
</ul>
<p>If we take a look at Table 1 with the DE results, it will become clearer how they did their analysis:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img_de/paper_table1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:80.0%"></p>
</figure>
</div>
<p>It looks like they performed pairwise comparisons between each combination of time and treatment. But this can in turn be done in two ways with DESeq2:</p>
<ul>
<li>After subsetting the dataset to each combination of time and treatment.</li>
<li>After creating a single independent variable that is a combination of time and treatment.</li>
</ul>
<p>The latter method is the more common one, and is what we will do below<a href="#fn3" class="footnote-ref" id="fnref3" role="doc-noteref"><sup>3</sup></a>.</p>
<hr style="height:1pt; visibility:hidden;">
</section>
<section id="setting-the-statistical-design" class="level3" data-number="5.2">
<h3 data-number="5.2" class="anchored" data-anchor-id="setting-the-statistical-design"><span class="header-section-number">5.2</span> Setting the statistical design</h3>
<p>As discussed above, we will now create a new variable that is a combination of <code>treatment</code> and <code>time</code>, and we’ll call it <code>group</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a combined variable called 'group':</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>dds<span class="sc">$</span>group <span class="ot">&lt;-</span> <span class="fu">factor</span>(<span class="fu">paste</span>(dds<span class="sc">$</span>treatment, dds<span class="sc">$</span>time, <span class="at">sep =</span> <span class="st">"_"</span>))</span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Which unique values does 'group' have, and how many samples are in each?</span></span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(dds<span class="sc">$</span>group)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
cathemerium_10dpi cathemerium_24hpi     control_10dpi     control_24hpi 
                4                 3                 4                 3 
   relictum_10dpi    relictum_24hpi 
                4                 4 </code></pre>
</div>
</div>
<p>Next, we set the analysis design:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="fu">design</span>(dds) <span class="ot">&lt;-</span> <span class="er">~</span> group</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we’re ready to run the DE analysis!</p>
<hr style="height:1pt; visibility:hidden;">
</section>
<section id="running-the-de-analysis" class="level3" data-number="5.3">
<h3 data-number="5.3" class="anchored" data-anchor-id="running-the-de-analysis"><span class="header-section-number">5.3</span> Running the DE analysis</h3>
<p>We perform the differential expression analysis with the <code>DEseq()</code> function:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeq</span>(dds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>estimating size factors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>estimating dispersions</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>gene-wise dispersion estimates</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>mean-dispersion relationship</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>final dispersion estimates</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>fitting model and testing</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>Note that we above, we assigned the output back to the <code>dds</code> object — in this case, the DE results are being <em>added</em> to the object!</p>
</div>
</div>
<p>The <code>DESeq()</code> function is a wrapper that performs three steps (functions) consecutively:</p>
<ul>
<li><code>estimateSizeFactors()</code> — “Normalization” by library size and composition.</li>
<li><code>estimateDispersions()</code> — Estimate gene-wise dispersion (variance in counts).</li>
<li><code>nbinomWaldTest(ddsObj)</code> — Fit the negative binomial GLM and calculate Wald statistics, which is the test statistic underlying the DE p-value.</li>
</ul>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
More about gene count normalization
</div>
</div>
<div class="callout-body-container callout-body">
<p>Note that <em>DESeq2</em> doesn’t actually normalize the counts in the sense that it produces a matrix with adjusted counts. Instead it uses raw counts and includes the size factors in the modeling. To learn more about gene count normalization, see <a href="https://www.youtube.com/watch?v=UFB993xufUU">this video</a> and <a href="https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html">this page</a>.</p>
</div>
</div>
<p><br></p>
</section>
</section>
<section id="extracting-the-de-results" class="level2" data-number="6">
<h2 data-number="6" class="anchored" data-anchor-id="extracting-the-de-results"><span class="header-section-number">6</span> Extracting the DE results</h2>
<p>A key thing to understand is that DESeq2 has automatically performed <strong>pairwise</strong> comparisons between <strong>each of the (6) levels</strong> of the <code>group</code> variable.</p>
<p>As such, we can extract a separate table with gene-wise results for each of these (15) pairwise comparisons. Or put differently, for any individual gene, it tested whether this gene is differentially expressed separately for each of the pairwise comparisons.</p>
<hr style="height:1pt; visibility:hidden;">
<section id="the-results-table" class="level3" data-number="6.1">
<h3 data-number="6.1" class="anchored" data-anchor-id="the-results-table"><span class="header-section-number">6.1</span> The results table</h3>
<p>We can extract the results for one pairwise comparison (which DESeq2 refers to as a <strong>contrast</strong>) at a time, by specifying it with the <code>contrast</code> argument as a vector of length 3:</p>
<ol type="1">
<li>The focal independent variable (here, <code>group</code>)</li>
<li>The first (reference) level of the independent variable (in the example below, <code>relictum_24hpi</code>)</li>
<li>The second level of the independent variable (in the example below, <code>control_24hpi</code>)</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>res_rc24 <span class="ot">&lt;-</span> <span class="fu">results</span>(dds, <span class="at">contrast =</span> <span class="fu">c</span>(<span class="st">"group"</span>, <span class="st">"relictum_24hpi"</span>, <span class="st">"control_24hpi"</span>))</span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(res_rc24)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>log2 fold change (MLE): group relictum_24hpi vs control_24hpi 
Wald test p-value: group relictum_24hpi vs control_24hpi 
DataFrame with 6 rows and 6 columns
       baseMean log2FoldChange     lfcSE      stat    pvalue      padj
      &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
ATP6  7658.0445      -0.416305  0.609133 -0.683438 0.4943300  0.776172
ATP8     4.9196      -1.311116  1.388811 -0.944057 0.3451406        NA
COX1 75166.8670      -0.590935  0.282075 -2.094958 0.0361747  0.208045
COX2  7807.1848      -0.610152  0.578401 -1.054893 0.2914743  0.615249
COX3 41037.7359      -0.400173  0.251760 -1.589498 0.1119479  0.388880
CYTB 36916.6130      -0.501653  0.261927 -1.915242 0.0554617  0.266528</code></pre>
</div>
</div>
<p>What do the columns in this table contain?</p>
<ul>
<li><strong><code>baseMean</code></strong>: Mean expression level across all samples.</li>
<li><strong><code>log2FoldChange</code></strong>: The “log2-fold change” of gene counts between the compared levels.</li>
<li><strong><code>lfcSE</code></strong>: The uncertainty in terms of the standard error (SE) of the log2-fold change estimate.</li>
<li><strong><code>stat</code></strong>: The value for the Wald test’s test statistic.</li>
<li><strong><code>pvalue</code></strong>: The <em>uncorrected</em> p-value from the Wald test.</li>
<li><strong><code>padj</code></strong>: The multiple-testing corrected p-value (i.e., adjusted p-value).</li>
</ul>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Log2-fold changes
</div>
</div>
<div class="callout-body-container callout-body">
<p>Log2-fold changes are the standard way of representing the <em>effect size</em> of expression level differences between two groups of interest.</p>
<p>A log2-fold change of 1 indicates that the expression in the reference level is two-fold <em>lower</em> than that of the other level, a log2-fold change of 2 indicates a four-fold difference, a log2-fold change of 3 indicates an eight-fold difference, and so on.</p>
<p>Similarly, <em>negative log2-fold</em> values indicate a change in gene counts in the other direction: the reference level is <em>higher</em> than the other level.</p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Multiple testing correction
</div>
</div>
<div class="callout-body-container callout-body">
<p>Because we are testing significance for <em>many</em> genes, we need to correct for multiple testing. DESeq2 uses the Benjamini-Hochberg False Discovery Rate (FDR) correction.</p>
<p>For more info, see: STATQUEST</p>
</div>
</div>
<p>How many adjusted p-values were less than 0.05?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(res_rc24<span class="sc">$</span>padj <span class="sc">&lt;</span> <span class="fl">0.05</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 801</code></pre>
</div>
</div>
<p>So, we have 801 Differentially Expressed Genes (<strong>DEG</strong>s) for this specific pairwise comparison.</p>
<p>How does that compare to what they found in the paper?</p>
<details>
<summary>
<strong>Your Turn:</strong> The paper’s Table 1 also reports numbers of up- and downregulated genes separately. Can you find this out for our DEGs? <em>(Click to see the solution)</em>
</summary>
</details>
<details>
<summary>
<strong>Your Turn:</strong> The paper’s Table 1 also reports the number of DEGs with an absolute log2-fold change greater than 1. Can you find this out for our DEGs? <em>(Click to see the solution)</em>
</summary>
</details>
<details>
<summary>
<strong>Your Turn:</strong> Extract the results for one or more other contrasts that are the paper’s Table 1, and compare the results.
</summary>
</details>
<details>
<summary>
<strong>Bonus section</strong>: <code>NA</code> values in the results table <em>(Click to expand)</em>
</summary>
<p>Some values in the results table can be set to <code>NA</code> for one of the following reasons:</p>
<ul>
<li><p>If a gene contains a sample with a count <strong>outlier</strong>, both the p-value and adjusted p-value will be set to <code>NA</code>. (DESeq2 performs outlier detection using Cook’s distance.)</p></li>
<li><p>If all samples have <strong>zero counts</strong> for a given gene, the <code>baseMean</code> column will be zero, and the log2-fold change estimates, p-value and adjusted p-value will all be set to <code>NA</code>.</p></li>
<li><p>DESeq2 also automatically filters genes with a <strong>low mean count</strong> in the sense that it does not include them in the multiple testing correction. Therefore, in such cases, the p-value will not be <code>NA</code>, but the <em>adjusted</em> p-value will be.</p>
<p>Because we have very low power to detect differential expression for such low-count genes, it is beneficial to remove them prior to the multiple testing correction: that way, the correction becomes less severe for the remaining genes.</p></li>
</ul>
<p>Let’s see how many genes have <code>NA</code> p-values:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Number of genes with NA p-value:</span></span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(<span class="fu">is.na</span>(res_rc24<span class="sc">$</span>pvalue))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1124</code></pre>
</div>
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="co"># As a proportion of the total number of genes in the test:</span></span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(<span class="fu">is.na</span>(res_rc24<span class="sc">$</span>pvalue)) <span class="sc">/</span> <span class="fu">nrow</span>(res_rc24)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.05961283</code></pre>
</div>
</div>
<p>And <code>NA</code> adjusted p-values:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Number of genes with NA p-value:</span></span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(<span class="fu">is.na</span>(res_rc24<span class="sc">$</span>padj))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 7283</code></pre>
</div>
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="co"># As a proportion of the total number of genes in the test:</span></span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(<span class="fu">is.na</span>(res_rc24<span class="sc">$</span>padj)) <span class="sc">/</span> <span class="fu">nrow</span>(res_rc24)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.3862636</code></pre>
</div>
</div>
</details>
<details>
<summary>
<strong>Bonus section</strong>: Exporting the results <em>(Click to expand)</em>
</summary>
<p>You may be wondering how we can save the DE results tables:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create the output directory, if necessary:</span></span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="st">"results/DE"</span>, <span class="at">recursive =</span> <span class="cn">TRUE</span>, <span class="at">showWarnings =</span> <span class="cn">FALSE</span>)</span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb73-4"><a href="#cb73-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Write the </span></span>
<span id="cb73-5"><a href="#cb73-5" aria-hidden="true" tabindex="-1"></a><span class="fu">write_tsv</span>(<span class="fu">as.data.frame</span>(res_rc24), <span class="st">"results/DE/resultsres_rc24.tsv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</details>
<p><br></p>
</section>
</section>
<section id="visualizing-the-de-results" class="level2" data-number="7">
<h2 data-number="7" class="anchored" data-anchor-id="visualizing-the-de-results"><span class="header-section-number">7</span> Visualizing the DE results</h2>
<p>We will create a few plots for the results for the <code>relictum_24hpi</code> vs.&nbsp;<code>control_24hpi</code> comparison, which we extracted above.</p>
<section id="volcano-plot" class="level3" data-number="7.1">
<h3 data-number="7.1" class="anchored" data-anchor-id="volcano-plot"><span class="header-section-number">7.1</span> Volcano plot</h3>
<p>For a nice overview of the results, we can plot a so-called “volcano plot” using the <code>EnhancedVolcano()</code> function from the package of the same name (<a href="https://bioconductor.org/packages/release/bioc/vignettes/EnhancedVolcano/inst/doc/EnhancedVolcano.html">see here for a “vignette” / tutorial</a>):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="fu">EnhancedVolcano</span>(</span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">toptable =</span> res_rc24,      <span class="co"># DESeq2 results to plot   </span></span>
<span id="cb74-3"><a href="#cb74-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">title =</span> <span class="st">"relictum vs. control at 24 hpi"</span>,</span>
<span id="cb74-4"><a href="#cb74-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> <span class="st">"log2FoldChange"</span>,     <span class="co"># Plot the log2-fold change along the x-axis</span></span>
<span id="cb74-5"><a href="#cb74-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">y =</span> <span class="st">"pvalue"</span>,             <span class="co"># Plot the p-value along the y-axis</span></span>
<span id="cb74-6"><a href="#cb74-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">lab =</span> <span class="fu">rownames</span>(res_rc24), <span class="co"># Use the rownames for the gene labels (though see below)</span></span>
<span id="cb74-7"><a href="#cb74-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">labSize =</span> <span class="dv">0</span>               <span class="co"># Omit gene labels</span></span>
<span id="cb74-8"><a href="#cb74-8" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<details>
<summary>
<em>Click to see the plot</em>
</summary>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="DE_files/figure-html/unnamed-chunk-45-1.png" class="img-fluid figure-img" style="width:85.0%"></p>
</figure>
</div>
</div>
</div>
</details>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Plot customization ideas
</div>
</div>
<div class="callout-body-container callout-body">
<p>Above, we turned the gene labeling off by setting <code>labSize = 0</code> — the default p-value cut-off for point labeling is <code>1e-5</code> (and the default log2-fold change cut-off is <code>1</code>): in this case, that would get quite busy with gene labels. We might want to try a plot with a stricter p-value cut-off that does show the gene labels.</p>
<p>Some further ideas for potential customization:</p>
<ul>
<li>We might want to get rid of the subtitle (“EnhancedVolcano”) and the caption (“total=…”).</li>
<li>Perhaps you want to exclude the outlier with the &gt;20 log2-fold change.</li>
<li>Conversely, you might want to <em>label</em> the abovementioned outlier.</li>
<li>You might want to try some other options you see in <a href="https://bioconductor.org/packages/release/bioc/vignettes/EnhancedVolcano/inst/doc/EnhancedVolcano.html">this vignette</a> or the help page which you can access by running <code>?EnhancedVolcano</code>.</li>
</ul>
</div>
</div>
<details>
<summary>
<strong>Your Turn:</strong> Consider some of the ideas above and customize the plot (<em>Click for an example</em>)
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a><span class="fu">EnhancedVolcano</span>(</span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">toptable =</span> res_rc24,      </span>
<span id="cb75-3"><a href="#cb75-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">title =</span> <span class="st">"relictum vs. control at 24 hpi"</span>,</span>
<span id="cb75-4"><a href="#cb75-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> <span class="st">"log2FoldChange"</span>,     </span>
<span id="cb75-5"><a href="#cb75-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">y =</span> <span class="st">"pvalue"</span>,             </span>
<span id="cb75-6"><a href="#cb75-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">lab =</span> <span class="fu">rownames</span>(res_rc24), </span>
<span id="cb75-7"><a href="#cb75-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">labSize =</span> <span class="dv">3</span>,               <span class="co"># Now we will show the gene labels</span></span>
<span id="cb75-8"><a href="#cb75-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">pCutoff =</span> <span class="fl">10e-10</span>,          <span class="co"># Modify the p-value cut-off</span></span>
<span id="cb75-9"><a href="#cb75-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">subtitle =</span> <span class="cn">NULL</span>,           <span class="co"># Remove the subtitle</span></span>
<span id="cb75-10"><a href="#cb75-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">caption =</span> <span class="cn">NULL</span>,            <span class="co"># Remove the caption</span></span>
<span id="cb75-11"><a href="#cb75-11" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="DE_files/figure-html/unnamed-chunk-46-1.png" class="img-fluid figure-img" style="width:85.0%"></p>
</figure>
</div>
</div>
</div>
<details>
<summary>
<strong>Your Turn:</strong> Figure out the identity of the abovementioned log2-fold change outlier either by labeling it in the plot, or by filtering the <code>res_rc24</code> table (<em>Click for the solution</em>)
</summary>
<ul>
<li>By labeling it in the plot:</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="fu">EnhancedVolcano</span>(</span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">toptable =</span> res_rc24,      </span>
<span id="cb76-3"><a href="#cb76-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">title =</span> <span class="st">"relictum vs. control at 24 hpi"</span>,</span>
<span id="cb76-4"><a href="#cb76-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> <span class="st">"log2FoldChange"</span>,     </span>
<span id="cb76-5"><a href="#cb76-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">y =</span> <span class="st">"pvalue"</span>,             </span>
<span id="cb76-6"><a href="#cb76-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">lab =</span> <span class="fu">rownames</span>(res_rc24), </span>
<span id="cb76-7"><a href="#cb76-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">labSize =</span> <span class="dv">3</span>,               </span>
<span id="cb76-8"><a href="#cb76-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">pCutoff =</span> <span class="fl">10e-4</span>,           <span class="co"># Modify the p-value cut-off</span></span>
<span id="cb76-9"><a href="#cb76-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">FCcutoff =</span> <span class="dv">20</span>,</span>
<span id="cb76-10"><a href="#cb76-10" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 1 rows containing missing values (`geom_vline()`).</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="DE_files/figure-html/unnamed-chunk-47-1.png" class="img-fluid figure-img" style="width:85.0%"></p>
</figure>
</div>
</div>
</div>
<ul>
<li>By filtering the results table:</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="fu">as.data.frame</span>(res_rc24) <span class="sc">|&gt;</span> <span class="fu">filter</span>(log2FoldChange <span class="sc">&gt;</span> <span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>              baseMean log2FoldChange    lfcSE     stat       pvalue
LOC120413430  7.540043       24.46898 5.397990 4.532979           NA
LOC120431476 39.720375       23.01445 5.301369 4.341228 1.416886e-05
                     padj
LOC120413430           NA
LOC120431476 0.0008584398</code></pre>
</div>
</div>
</details>
<hr style="height:1pt; visibility:hidden;">
</details></section>
<section id="plot-specific-genes" class="level3" data-number="7.2">
<h3 data-number="7.2" class="anchored" data-anchor-id="plot-specific-genes"><span class="header-section-number">7.2</span> Plot specific genes</h3>
<p>We can also create plots of expression levels for individual genes. That is especially interesting for genes with highly significant differential expression. So let’s plot the most highly significant DEG.</p>
<p>First, let’s create a vector with most highly signifant DEGs, which we’ll use again for the heatmap below.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>top25_DE <span class="ot">&lt;-</span> <span class="fu">row.names</span>(res_rc24[<span class="fu">order</span>(res_rc24<span class="sc">$</span>padj)[<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>], ])</span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-3"><a href="#cb80-3" aria-hidden="true" tabindex="-1"></a>top25_DE</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "LOC120423768" "LOC120423767" "LOC120414587" "LOC128092307" "LOC120431154"
 [6] "LOC120427827" "LOC120415152" "LOC120422735" "LOC120431739" "LOC120431733"
[11] "LOC120428214" "LOC120427588" "LOC120415540" "LOC120415522" "LOC120429000"
[16] "LOC120414889" "LOC120413491" "LOC120414802" "LOC120423826" "LOC120429211"</code></pre>
</div>
</div>
<p>DESeq2 has a plotting function but the plot is not very good. We will still use that function to quickly extract the counts for our gene of interest in the right format for plotting, using <code>returnData = TRUE</code> (which will also make it omit the plot):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a>focal_gene_counts <span class="ot">&lt;-</span> <span class="fu">plotCounts</span>(</span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a>  dds,</span>
<span id="cb82-3"><a href="#cb82-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene =</span> top25_DE[<span class="dv">1</span>],</span>
<span id="cb82-4"><a href="#cb82-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">intgroup =</span> <span class="fu">c</span>(<span class="st">"time"</span>, <span class="st">"treatment"</span>),</span>
<span id="cb82-5"><a href="#cb82-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">returnData =</span> <span class="cn">TRUE</span></span>
<span id="cb82-6"><a href="#cb82-6" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(focal_gene_counts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<details>
<summary>
<em>Click to see the result</em>
</summary>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>                 count  time   treatment
ERR10802863 1543.81532 24hpi     control
ERR10802864 2279.03704 24hpi cathemerium
ERR10802865   25.42295 24hpi    relictum
ERR10802866 1105.75009 24hpi     control
ERR10802867 1199.28425 24hpi cathemerium
ERR10802868   32.14394 24hpi    relictum</code></pre>
</div>
</div>
</details>
<p>Now we can make the plot:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(focal_gene_counts, <span class="fu">aes</span>(<span class="at">x =</span> treatment, <span class="at">y =</span> count, <span class="at">fill =</span> treatment)) <span class="sc">+</span></span>
<span id="cb85-2"><a href="#cb85-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">alpha =</span> <span class="fl">0.5</span>, <span class="at">outlier.shape =</span> <span class="cn">NA</span>) <span class="sc">+</span></span>
<span id="cb85-3"><a href="#cb85-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">4</span>, <span class="at">shape =</span> <span class="dv">21</span>, <span class="at">position =</span> <span class="fu">position_jitter</span>(<span class="at">w =</span> <span class="fl">0.1</span>, <span class="at">h =</span> <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb85-4"><a href="#cb85-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="fu">vars</span>(time)) <span class="sc">+</span></span>
<span id="cb85-5"><a href="#cb85-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb85-6"><a href="#cb85-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<details>
<summary>
<em>Click to see the plot</em>
</summary>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="DE_files/figure-html/unnamed-chunk-54-1.png" class="img-fluid figure-img" style="width:85.0%"></p>
</figure>
</div>
</div>
</div>
</details>
<details>
<summary>
<strong>Your Turn:</strong> Plot the gene with the very high log-fold change value that we saw when making the volcano plot. <em>(Click for the solution)</em>
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a>focal_gene_counts <span class="ot">&lt;-</span> <span class="fu">plotCounts</span>(</span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a>  dds,</span>
<span id="cb86-3"><a href="#cb86-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene =</span> <span class="st">"LOC120431476"</span>,</span>
<span id="cb86-4"><a href="#cb86-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">intgroup =</span> <span class="fu">c</span>(<span class="st">"time"</span>, <span class="st">"treatment"</span>),</span>
<span id="cb86-5"><a href="#cb86-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">returnData =</span> <span class="cn">TRUE</span></span>
<span id="cb86-6"><a href="#cb86-6" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb86-7"><a href="#cb86-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-8"><a href="#cb86-8" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(focal_gene_counts, <span class="fu">aes</span>(<span class="at">x =</span> treatment, <span class="at">y =</span> count, <span class="at">fill =</span> treatment)) <span class="sc">+</span></span>
<span id="cb86-9"><a href="#cb86-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">alpha =</span> <span class="fl">0.5</span>, <span class="at">outlier.shape =</span> <span class="cn">NA</span>) <span class="sc">+</span></span>
<span id="cb86-10"><a href="#cb86-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">4</span>, <span class="at">shape =</span> <span class="dv">21</span>, <span class="at">position =</span> <span class="fu">position_jitter</span>(<span class="at">w =</span> <span class="fl">0.1</span>, <span class="at">h =</span> <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb86-11"><a href="#cb86-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="fu">vars</span>(time)) <span class="sc">+</span></span>
<span id="cb86-12"><a href="#cb86-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb86-13"><a href="#cb86-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="DE_files/figure-html/unnamed-chunk-55-1.png" class="img-fluid figure-img" style="width:85.0%"></p>
</figure>
</div>
</div>
</div>
</details>
<details>
<summary>
<strong>Bonus Exercise:</strong> Plot one or a few more of the top-DE genes. Do they have similar expression patterns across treatment and timepoints as the first one?
</summary>
</details>
<hr style="height:1pt; visibility:hidden;">
</section>
<section id="heatmaps" class="level3" data-number="7.3">
<h3 data-number="7.3" class="anchored" data-anchor-id="heatmaps"><span class="header-section-number">7.3</span> Heatmaps</h3>
<p>We can create heatmaps with the <code>pheatmap</code> function. Let’s start by creating a function that will plot a heatmap given a vector of gene IDs and a DESeq2 object <code>dds</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a>norm_mat <span class="ot">&lt;-</span> <span class="fu">assay</span>(dds_vst)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a>norm_mat_sel <span class="ot">&lt;-</span> norm_mat[<span class="fu">match</span>(top25_DE, <span class="fu">rownames</span>(norm_mat)), ]</span>
<span id="cb88-2"><a href="#cb88-2" aria-hidden="true" tabindex="-1"></a>meta_sort <span class="ot">&lt;-</span> meta <span class="sc">|&gt;</span></span>
<span id="cb88-3"><a href="#cb88-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(treatment, time) <span class="sc">|&gt;</span></span>
<span id="cb88-4"><a href="#cb88-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(treatment, time)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(norm_mat_sel,</span>
<span id="cb89-2"><a href="#cb89-2" aria-hidden="true" tabindex="-1"></a>         <span class="at">cluster_rows =</span> <span class="cn">TRUE</span>,</span>
<span id="cb89-3"><a href="#cb89-3" aria-hidden="true" tabindex="-1"></a>         <span class="at">cluster_cols =</span> <span class="cn">FALSE</span>,</span>
<span id="cb89-4"><a href="#cb89-4" aria-hidden="true" tabindex="-1"></a>         <span class="at">show_rownames =</span> <span class="cn">FALSE</span>,</span>
<span id="cb89-5"><a href="#cb89-5" aria-hidden="true" tabindex="-1"></a>         <span class="at">annotation_col =</span> meta_sort,</span>
<span id="cb89-6"><a href="#cb89-6" aria-hidden="true" tabindex="-1"></a>         <span class="at">scale =</span> <span class="st">"row"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<details>
<summary>
<em>Click to see the plot</em>
</summary>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="DE_files/figure-html/unnamed-chunk-59-1.png" class="img-fluid figure-img" style="width:85.0%"></p>
</figure>
</div>
</div>
</div>
</details>
<details>
<summary>
<strong>Your Turn:</strong> Make a heatmap with the top-25 most-highly expressed genes <em>(Click for a hint: how to get that top-25)</em>
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a>top25_hi <span class="ot">&lt;-</span> <span class="fu">names</span>(<span class="fu">sort</span>(<span class="fu">rowMeans</span>(norm_mat), <span class="at">decreasing =</span> <span class="cn">TRUE</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</details>
<details>
<summary>
<em>Click for the solution</em>
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a>norm_mat_sel <span class="ot">&lt;-</span> norm_mat[<span class="fu">match</span>(top25_hi, <span class="fu">rownames</span>(norm_mat)), ]</span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a>meta_sort <span class="ot">&lt;-</span> meta <span class="sc">|&gt;</span></span>
<span id="cb91-3"><a href="#cb91-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(treatment, time) <span class="sc">|&gt;</span></span>
<span id="cb91-4"><a href="#cb91-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(treatment, time)</span>
<span id="cb91-5"><a href="#cb91-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb91-6"><a href="#cb91-6" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(norm_mat_sel,</span>
<span id="cb91-7"><a href="#cb91-7" aria-hidden="true" tabindex="-1"></a>         <span class="at">cluster_rows =</span> <span class="cn">TRUE</span>,</span>
<span id="cb91-8"><a href="#cb91-8" aria-hidden="true" tabindex="-1"></a>         <span class="at">cluster_cols =</span> <span class="cn">FALSE</span>,</span>
<span id="cb91-9"><a href="#cb91-9" aria-hidden="true" tabindex="-1"></a>         <span class="at">show_rownames =</span> <span class="cn">FALSE</span>,</span>
<span id="cb91-10"><a href="#cb91-10" aria-hidden="true" tabindex="-1"></a>         <span class="at">annotation_col =</span> meta_sort,</span>
<span id="cb91-11"><a href="#cb91-11" aria-hidden="true" tabindex="-1"></a>         <span class="at">scale =</span> <span class="st">"row"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="DE_files/figure-html/unnamed-chunk-61-1.png" class="img-fluid figure-img" style="width:85.0%"></p>
</figure>
</div>
</div>
</div>
</details>
<hr style="height:1pt; visibility:hidden;">
<p><br></p>
</section>
</section>
<section id="in-closing" class="level2" data-number="8">
<h2 data-number="8" class="anchored" data-anchor-id="in-closing"><span class="header-section-number">8</span> In Closing</h2>
<section id="next-steps" class="level3" data-number="8.1">
<h3 data-number="8.1" class="anchored" data-anchor-id="next-steps"><span class="header-section-number">8.1</span> Next steps</h3>
<p>Some potential next steps for this analysis are:</p>
<ul>
<li>Run the DE analysis for all pairwise comparisons</li>
<li>Compare DEGs across pairwise (this would allow us to make the upset plot in Figure 2 of the paper)</li>
<li>Perm</li>
<li>Enrichment analysis</li>
</ul>
<p><br></p>


</section>
</section>


<a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a><div id="quarto-appendix" class="default"><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><h2 class="anchored quarto-appendix-heading">Footnotes</h2>

<ol>
<li id="fn1"><p>And even for more basic tasks, it is common to use packages that are preferred over the functionality that is by default available in R, like in the case of plotting.<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn2"><p>Yes, this may be surprising!<a href="#fnref2" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn3"><p>I can’t tell from the paper which method they used<a href="#fnref3" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>