<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Jelmer Poelstra">
<meta name="dcterms.date" content="2024-02-02">

<title>ENTMLGY 6703 - Jelmer’s lectures - Week 4 lab: RNA-seq differential expression analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">ENTMLGY 6703 - Jelmer’s lectures</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html" rel="" target=""><i class="bi bi-house" role="img">
</i> 
 <span class="menu-text"> </span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-week-3-hts" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Week 3: HTS</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-week-3-hts">    
        <li>
    <a class="dropdown-item" href="../lectures/sequencing.html" rel="" target="">
 <span class="dropdown-text">Lecture: High-throughput Sequencing &amp; Genomes</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../labs/infrastructure.html" rel="" target="">
 <span class="dropdown-text">Lab part 1: Computational Infrastructure</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../labs/data.html" rel="" target="">
 <span class="dropdown-text">Lab part 2: Working with genome and HTS Data</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-week-4-rna-seq" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Week 4: RNA-seq</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-week-4-rna-seq">    
        <li class="dropdown-header">Lecture: RNA-seq</li>
        <li class="dropdown-header">Lab: Differential expression analysis</li>
    </ul>
  </li>
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../about.html" rel="" target="">
 <span class="menu-text">About</span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label=""><i class="bi bi-github"></i></a>
      <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="quarto-navigation-tool-dropdown-0">
          <li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://github.com/jelmerp/ENTMLGY6703">
            Source Code
            </a>
          </li>
          <li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://github.com/jelmerp/ENTMLGY6703/issues/new">
            Report an Issue
            </a>
          </li>
      </ul>
    </div>
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction"><span class="header-section-number">1</span> Introduction</a></li>
  <li><a href="#getting-set-up" id="toc-getting-set-up" class="nav-link" data-scroll-target="#getting-set-up"><span class="header-section-number">2</span> Getting set up</a>
  <ul class="collapse">
  <li><a href="#start-an-rstudio-session-at-osc" id="toc-start-an-rstudio-session-at-osc" class="nav-link" data-scroll-target="#start-an-rstudio-session-at-osc"><span class="header-section-number">2.1</span> Start an RStudio session at OSC</a></li>
  <li><a href="#rstudio-settings-project" id="toc-rstudio-settings-project" class="nav-link" data-scroll-target="#rstudio-settings-project"><span class="header-section-number">2.2</span> RStudio settings &amp; Project</a></li>
  <li><a href="#load-the-necessary-packages" id="toc-load-the-necessary-packages" class="nav-link" data-scroll-target="#load-the-necessary-packages"><span class="header-section-number">2.3</span> Load the necessary packages</a></li>
  <li><a href="#define-our-inputs-and-outputs" id="toc-define-our-inputs-and-outputs" class="nav-link" data-scroll-target="#define-our-inputs-and-outputs"><span class="header-section-number">2.4</span> Define our inputs and outputs</a></li>
  <li><a href="#load-input-data" id="toc-load-input-data" class="nav-link" data-scroll-target="#load-input-data"><span class="header-section-number">2.5</span> Load input data</a></li>
  </ul></li>
  <li><a href="#create-a-deseq2-object-from-the-input-data" id="toc-create-a-deseq2-object-from-the-input-data" class="nav-link" data-scroll-target="#create-a-deseq2-object-from-the-input-data"><span class="header-section-number">3</span> Create a DESeq2 object from the input data</a></li>
  <li><a href="#explore-the-count-data" id="toc-explore-the-count-data" class="nav-link" data-scroll-target="#explore-the-count-data"><span class="header-section-number">4</span> Explore the count data</a>
  <ul class="collapse">
  <li><a href="#histogram-of-gene-counts" id="toc-histogram-of-gene-counts" class="nav-link" data-scroll-target="#histogram-of-gene-counts"><span class="header-section-number">4.1</span> Histogram of gene counts</a></li>
  </ul></li>
  <li><a href="#principal-component-analysis-pca" id="toc-principal-component-analysis-pca" class="nav-link" data-scroll-target="#principal-component-analysis-pca"><span class="header-section-number">5</span> Principal Component Analysis (PCA)</a>
  <ul class="collapse">
  <li><a href="#run-the-pca-and-prepare-for-plotting" id="toc-run-the-pca-and-prepare-for-plotting" class="nav-link" data-scroll-target="#run-the-pca-and-prepare-for-plotting"><span class="header-section-number">5.1</span> Run the PCA and prepare for plotting</a></li>
  <li><a href="#plot-the-pca-results" id="toc-plot-the-pca-results" class="nav-link" data-scroll-target="#plot-the-pca-results"><span class="header-section-number">5.2</span> Plot the PCA results</a></li>
  <li><a href="#plot-again-with-sample-names" id="toc-plot-again-with-sample-names" class="nav-link" data-scroll-target="#plot-again-with-sample-names"><span class="header-section-number">5.3</span> Plot again – with sample names</a></li>
  </ul></li>
  <li><a href="#de-analysis-full-dataset" id="toc-de-analysis-full-dataset" class="nav-link" data-scroll-target="#de-analysis-full-dataset"><span class="header-section-number">6</span> DE analysis – full dataset</a>
  <ul class="collapse">
  <li><a href="#the-results-table" id="toc-the-results-table" class="nav-link" data-scroll-target="#the-results-table"><span class="header-section-number">6.1</span> The results table</a></li>
  <li><a href="#na-values-in-the-results-table" id="toc-na-values-in-the-results-table" class="nav-link" data-scroll-target="#na-values-in-the-results-table"><span class="header-section-number">6.2</span> <code>NA</code> values in the results table</a></li>
  </ul></li>
  <li><a href="#de-analysis-contrast-two-custom-groups" id="toc-de-analysis-contrast-two-custom-groups" class="nav-link" data-scroll-target="#de-analysis-contrast-two-custom-groups"><span class="header-section-number">7</span> DE analysis – contrast two custom groups</a></li>
  <li><a href="#visually-exploring-the-results" id="toc-visually-exploring-the-results" class="nav-link" data-scroll-target="#visually-exploring-the-results"><span class="header-section-number">8</span> Visually exploring the results</a>
  <ul class="collapse">
  <li><a href="#volcano-plot" id="toc-volcano-plot" class="nav-link" data-scroll-target="#volcano-plot"><span class="header-section-number">8.1</span> Volcano plot</a></li>
  <li><a href="#plot-specific-genes" id="toc-plot-specific-genes" class="nav-link" data-scroll-target="#plot-specific-genes"><span class="header-section-number">8.2</span> Plot specific genes</a></li>
  <li><a href="#heatmap" id="toc-heatmap" class="nav-link" data-scroll-target="#heatmap"><span class="header-section-number">8.3</span> Heatmap</a></li>
  <li><a href="#export-the-results" id="toc-export-the-results" class="nav-link" data-scroll-target="#export-the-results"><span class="header-section-number">8.4</span> Export the results</a></li>
  </ul></li>
  <li><a href="#de-analysis-with-two-factors" id="toc-de-analysis-with-two-factors" class="nav-link" data-scroll-target="#de-analysis-with-two-factors"><span class="header-section-number">9</span> DE analysis – with two factors</a>
  <ul class="collapse">
  <li><a href="#controlling-for-one-factor" id="toc-controlling-for-one-factor" class="nav-link" data-scroll-target="#controlling-for-one-factor"><span class="header-section-number">9.1</span> Controlling for one factor</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Week 4 lab: RNA-seq differential expression analysis</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Jelmer Poelstra </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">February 2, 2024</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<section id="introduction" class="level2" data-number="1">
<h2 data-number="1" class="anchored" data-anchor-id="introduction"><span class="header-section-number">1</span> Introduction</h2>
<p>For differential expression analysis, we will be using the popular <strong>DESeq2</strong> R/Bioconductor package (<a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8">paper</a>, <a href="https://bioconductor.org/packages/release/bioc/html/DESeq2.html">website</a>).</p>
<p><br></p>
</section>
<section id="getting-set-up" class="level2" data-number="2">
<h2 data-number="2" class="anchored" data-anchor-id="getting-set-up"><span class="header-section-number">2</span> Getting set up</h2>
<section id="start-an-rstudio-session-at-osc" class="level3" data-number="2.1">
<h3 data-number="2.1" class="anchored" data-anchor-id="start-an-rstudio-session-at-osc"><span class="header-section-number">2.1</span> Start an RStudio session at OSC</h3>
<details>
<summary>
Click to see the instructions
</summary>
<ol type="1">
<li>Log in to OSC at <a href="https://ondemand.osc.edu" class="uri">https://ondemand.osc.edu</a>.</li>
<li>Click on <code>Interactive Apps</code> (top bar) &gt; <code>RStudio Server</code> (all the way at the bottom)</li>
<li>Fill out the form, selecting (also shown in the figure below):
<ul>
<li>Cluster: <code>Pitzer</code></li>
<li>R version: <code>4.3.0</code></li>
<li>Project: <code>PAS2250</code></li>
<li>Number of hours: <code>4</code></li>
<li>Node type: <code>any</code></li>
<li>Number of cores: <code>2</code></li>
</ul></li>
<li>Click the <code>Launch</code> button</li>
</ol>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img_DE/rstudio_form.png" class="img-fluid figure-img" style="width:50.0%"></p>
</figure>
</div>
<ul>
<li>Now, you should see a box like this:</li>
</ul>
<p align="center">
<img src="img_DE/osc_queued.png" width="700">
</p>
<ul>
<li>Your job should start running pretty soon, and when it’s ready the box should look like this:</li>
</ul>
<p align="center">
<img src="img_DE/osc_running.png" width="700">
</p>
<ul>
<li>Click <code>Connect to RStudio Server</code> at the bottom of the box, and an RStudio Server instance will open. You’re ready to go!</li>
</ul>
</details>
</section>
<section id="rstudio-settings-project" class="level3" data-number="2.2">
<h3 data-number="2.2" class="anchored" data-anchor-id="rstudio-settings-project"><span class="header-section-number">2.2</span> RStudio settings &amp; Project</h3>
<section id="rstudio-settings" class="level4">
<h4 class="anchored" data-anchor-id="rstudio-settings">RStudio settings</h4>
<ul>
<li>Base R pipe</li>
<li>Don’t save the environment</li>
</ul>
</section>
<section id="rstudio-project" class="level4">
<h4 class="anchored" data-anchor-id="rstudio-project">RStudio Project</h4>
<p>Using an “RStudio Project” will most of all help with making sure your working directory in R is correct.</p>
<p>Click <code>File</code> &gt; <code>New Project</code> &gt; <code>Existing Directory</code> (in the popup) &gt; <code>Browse...</code> to browse to your personal dir in <code>/fs/scratch/PAS2250/ENT6703</code> &gt; <code>Create Project</code>.</p>
</section>
</section>
<section id="load-the-necessary-packages" class="level3" data-number="2.3">
<h3 data-number="2.3" class="anchored" data-anchor-id="load-the-necessary-packages"><span class="header-section-number">2.3</span> Load the necessary packages</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span> <span class="st">"pacman"</span> <span class="sc">%in%</span> <span class="fu">installed.packages</span>()) <span class="fu">install.packages</span>(<span class="st">"pacman"</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>packages <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"DESeq2"</span>,          <span class="co"># Differential expression analysis</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>              <span class="st">"tidyverse"</span>,       <span class="co"># Misc. data manipulation and plotting</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>              <span class="st">"ggrepel"</span>,         <span class="co"># PCA with sample IDs</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>              <span class="st">"pheatmap"</span>)        <span class="co"># Heatmap plot</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>pacman<span class="sc">::</span><span class="fu">p_load</span>(<span class="at">char =</span> packages)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">theme_set</span>(<span class="fu">theme_bw</span>())  <span class="co"># Set ggplot theme</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="define-our-inputs-and-outputs" class="level3" data-number="2.4">
<h3 data-number="2.4" class="anchored" data-anchor-id="define-our-inputs-and-outputs"><span class="header-section-number">2.4</span> Define our inputs and outputs</h3>
<p>For the differential expression analysis, we have the following input files:</p>
<ul>
<li><strong>Gene counts table</strong> – produced by running the <a href="https://nf-co.re/rnaseq">nf-core rnaseq workflow</a> on the input data we saw last week</li>
<li><strong>Metadata table</strong> – The metadata we saw last week, to enable between-treatment comparisons</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>count_table_file <span class="ot">&lt;-</span> <span class="st">"results/counts/salmon.merged.gene_counts_length_scaled.tsv"</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>metadata_file <span class="ot">&lt;-</span> <span class="st">"data/meta/metadata.txt"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The output directory – and we create it if it doesn’t already exist:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>outdir <span class="ot">&lt;-</span> <span class="st">"results/DE"</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(outdir, <span class="at">recursive =</span> <span class="cn">TRUE</span>, <span class="at">showWarnings =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="load-input-data" class="level3" data-number="2.5">
<h3 data-number="2.5" class="anchored" data-anchor-id="load-input-data"><span class="header-section-number">2.5</span> Load input data</h3>
<p>Load the count table:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>count_mat <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(mat_file, <span class="at">show_col_types =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Load the metadata information:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>meta <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(meta_file, <span class="at">show_col_types =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
</section>
</section>
<section id="create-a-deseq2-object-from-the-input-data" class="level2" data-number="3">
<h2 data-number="3" class="anchored" data-anchor-id="create-a-deseq2-object-from-the-input-data"><span class="header-section-number">3</span> Create a DESeq2 object from the input data</h2>
<p>Prepare the count table:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>count_mat <span class="ot">&lt;-</span> count_mat <span class="sc">|&gt;</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">column_to_rownames</span>(<span class="st">"gene_id"</span>) <span class="sc">|&gt;</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>gene_name) <span class="sc">|&gt;</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">round</span>() <span class="sc">|&gt;</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.matrix</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Prepare the metadata:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>meta <span class="ot">&lt;-</span> meta <span class="sc">|&gt;</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(sample_id) <span class="sc">|&gt;</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">column_to_rownames</span>(<span class="st">"sample_id"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>When creating the DESeq2 object, DESeq2 assumes that sample IDs in both tables match and are provided in the same order. Let’s make sure this is indeed the case:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">all</span>(<span class="fu">row.names</span>(meta) <span class="sc">==</span> <span class="fu">colnames</span>(count_mat))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We will create the DESeq2 object using the function <code>DESeqDataSetFromMatrix()</code>, which we will provide with three pieces of information:</p>
<ul>
<li>The count data with argument <code>countData</code>.</li>
<li>The metadata with argument <code>colData</code>.</li>
<li>The model design for the DE analysis – argument <code>design</code>.<br>
For now, we will specify <code>~1</code>, which means “no design” – we will change this before the actual DE analysis.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span>(</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">countData =</span> count_mat,</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">colData =</span> meta,</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">design =</span> <span class="sc">~</span> <span class="dv">1</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
</section>
<section id="explore-the-count-data" class="level2" data-number="4">
<h2 data-number="4" class="anchored" data-anchor-id="explore-the-count-data"><span class="header-section-number">4</span> Explore the count data</h2>
<p>What are number of rows and columns of the count matrix?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(counts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>How many genes have non-zero counts?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(counts[<span class="fu">rowSums</span>(counts) <span class="sc">&gt;</span> <span class="dv">0</span>, ])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>How many genes have total counts of at least 10?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(counts[<span class="fu">rowSums</span>(counts) <span class="sc">&gt;=</span> <span class="dv">10</span>, ])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="histogram-of-gene-counts" class="level3" data-number="4.1">
<h3 data-number="4.1" class="anchored" data-anchor-id="histogram-of-gene-counts"><span class="header-section-number">4.1</span> Histogram of gene counts</h3>
<p>Let’s plot a histogram of gene counts:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">theme_set</span>(<span class="fu">theme_bw</span>())</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>summed_gene_counts <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">gene_count =</span> <span class="fu">rowSums</span>(counts)) <span class="sc">%&gt;%</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="st">"gene_id"</span>)</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> summed_gene_counts) <span class="sc">+</span></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="fu">aes</span>(<span class="at">x =</span> gene_count), <span class="at">binwidth =</span> <span class="dv">10000</span>) <span class="sc">+</span></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_log10</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">0</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Zoom in a bit:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> summed_gene_counts) <span class="sc">+</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="fu">aes</span>(<span class="at">x =</span> gene_count), <span class="at">binwidth =</span> <span class="dv">1000</span>) <span class="sc">+</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_log10</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">200000</span>), <span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.margin =</span> <span class="fu">margin</span>(<span class="fl">0.5</span>, <span class="fl">0.7</span>, <span class="fl">0.5</span>, <span class="fl">0.5</span>, <span class="st">"cm"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>How are counts distributed across samples? That is, we would like a sum of counts for each column. To get this, we use the <code>apply()</code> function, which can apply a function (in our case <code>sum()</code>) to all columns (hence <code>MARGIN = 2</code> – for rows, use <code>1</code>) of our <code>counts</code> dataframe:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">apply</span>(<span class="at">X =</span> counts, <span class="at">MARGIN =</span> <span class="dv">2</span>, <span class="at">FUN =</span> sum)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
</section>
</section>
<section id="principal-component-analysis-pca" class="level2" data-number="5">
<h2 data-number="5" class="anchored" data-anchor-id="principal-component-analysis-pca"><span class="header-section-number">5</span> Principal Component Analysis (PCA)</h2>
<section id="run-the-pca-and-prepare-for-plotting" class="level3" data-number="5.1">
<h3 data-number="5.1" class="anchored" data-anchor-id="run-the-pca-and-prepare-for-plotting"><span class="header-section-number">5.1</span> Run the PCA and prepare for plotting</h3>
<p>First, we normalize the count data to have even sampling across samples (with respect to library size) and approximately even variance:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>vsd <span class="ot">&lt;-</span> <span class="fu">varianceStabilizingTransformation</span>(dds_raw, <span class="at">blind =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Next, we run the PCA and retrieve the data to plot with <em>ggplot2</em>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>pcaData <span class="ot">&lt;-</span> <span class="fu">plotPCA</span>(vsd,</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>                   <span class="at">ntop =</span> <span class="dv">500</span>,</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>                   <span class="at">intgroup =</span> <span class="fu">c</span>(<span class="st">"AMF"</span>, <span class="st">"Treatment"</span>),</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>                   <span class="at">returnData =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We extract the percentage of variance explained by different principal components, so we can later add this information to the plot:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>percentVar <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="dv">100</span> <span class="sc">*</span> <span class="fu">attr</span>(pcaData, <span class="st">"percentVar"</span>))</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>percentVar</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="plot-the-pca-results" class="level3" data-number="5.2">
<h3 data-number="5.2" class="anchored" data-anchor-id="plot-the-pca-results"><span class="header-section-number">5.2</span> Plot the PCA results</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(pcaData,</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> PC1, <span class="at">y =</span> PC2, <span class="at">color =</span> AMF, <span class="at">shape =</span> Treatment)) <span class="sc">+</span></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">6</span>) <span class="sc">+</span></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="fu">paste0</span>(<span class="st">"PC1: "</span>, percentVar[<span class="dv">1</span>], <span class="st">"% of variance"</span>)) <span class="sc">+</span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="fu">paste0</span>(<span class="st">"PC2: "</span>, percentVar[<span class="dv">2</span>], <span class="st">"% of variance"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="plot-again-with-sample-names" class="level3" data-number="5.3">
<h3 data-number="5.3" class="anchored" data-anchor-id="plot-again-with-sample-names"><span class="header-section-number">5.3</span> Plot again – with sample names</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(pcaData,</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(PC1, PC2, <span class="at">color =</span> AMF, <span class="at">shape =</span> Treatment)) <span class="sc">+</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_label_repel</span>(<span class="fu">aes</span>(<span class="at">label =</span> name)) <span class="sc">+</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="fu">paste0</span>(<span class="st">"PC1: "</span>, percentVar[<span class="dv">1</span>], <span class="st">"% of variance"</span>)) <span class="sc">+</span></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="fu">paste0</span>(<span class="st">"PC2: "</span>, percentVar[<span class="dv">2</span>], <span class="st">"% of variance"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
</section>
</section>
<section id="de-analysis-full-dataset" class="level2" data-number="6">
<h2 data-number="6" class="anchored" data-anchor-id="de-analysis-full-dataset"><span class="header-section-number">6</span> DE analysis – full dataset</h2>
<p>The design has two factors: <code>AMF</code> and <code>Treatment</code>. Rather than fit a multivariate model, we can start by merging the two into a single factor called <code>group</code>, and fit a univariate model with this factor.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>dds_raw<span class="sc">$</span>group <span class="ot">&lt;-</span> <span class="fu">factor</span>(<span class="fu">paste</span>(dds_raw<span class="sc">$</span>AMF, dds_raw<span class="sc">$</span>Treatment, <span class="at">sep =</span> <span class="st">"_"</span>))</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(dds_raw<span class="sc">$</span>group)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We will set the “reference” level of the factor to be the double negative control (empty substrate, no Agrobacteria):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>dds_raw<span class="sc">$</span>group <span class="ot">&lt;-</span> <span class="fu">relevel</span>(dds_raw<span class="sc">$</span>group, <span class="at">ref =</span> <span class="st">"control_mock"</span>)</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>dds_raw<span class="sc">$</span>group</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Next, we set the analysis design:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">design</span>(dds_raw) <span class="ot">&lt;-</span> <span class="fu">formula</span>(<span class="sc">~</span> group)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>And finally, we perform the differential expression analysis with the <code>DEseq()</code> function:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeq</span>(dds_raw)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code>DESeq()</code> function above performs three steps consecutively:</p>
<ul>
<li><p><code>estimateSizeFactors()</code> – “Normalization” by library size and composition.</p>
<p>Note that <em>DESeq2</em> doesn’t actually normalize the counts in the sense that it produces a matrix with adjusted counts. Instead it uses raw counts and includes the size factors in the modeling.</p>
<p>To learn more about gene count normalization, see <a href="https://www.youtube.com/watch?v=UFB993xufUU">this video</a> and <a href="https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html">this page</a>.</p></li>
<li><p><code>estimateDispersions()</code> – Estimate gene-wise dispersion (variance in counts).</p></li>
<li><p><code>nbinomWaldTest(ddsObj)</code> – Fit the negative binomial GLM and calculate Wald statistics, which is the test statistic underlying the p-value for whether a gene is differentially expressed.</p></li>
</ul>
<p>These functions could also be called separately, which would be useful if you want to be able to change more defaults.</p>
<section id="the-results-table" class="level3" data-number="6.1">
<h3 data-number="6.1" class="anchored" data-anchor-id="the-results-table"><span class="header-section-number">6.1</span> The results table</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">results</span>(dds)</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(res)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>By default, the results table prints statistics <strong>comparing the last level of the factor with the first level</strong>: that is, log-fold change and p-values describe differences between these two levels specifically. However, we can easily extract equivalent statistics for any pairwise comparison among our factor levels, which we will see later.</p>
<p>For now, we will explore what each column in this table means:</p>
<ul>
<li><p>The <strong><code>baseMean</code></strong> column contains the mean expression level across all samples.</p></li>
<li><p>The <strong><code>log2FoldChange</code></strong> column contains the log2-fold change of gene counts between the compared levels, that is, it represents the <em>effect size</em>.</p>
<p>A log2-fold change of 1 indicates that the expression in the reference level is two-fold <em>lower</em> than that of the other level, a log2-fold change of 2 indicates a four-fold difference, a log2-fold change of 3 indicates an eight-fold difference, and so on.</p>
<p>Similarly, <em>negative log2-fold</em> values indicate a change in gene counts in the other direction: the reference level is <em>higher</em> than the other level.</p></li>
<li><p>The <code>lfcSE</code> column indicates the uncertainty in terms of the standard error (SE) of the log2-fold change estimate.</p></li>
<li><p>The <strong><code>stat</code></strong> column indicates the value for the Wald test’s test statistic.</p></li>
<li><p>The <strong><code>pvalue</code></strong> column reported the <em>uncorrected</em> p-value from the Wald test.</p></li>
<li><p>Because we are testing significance for <em>many</em> genes, we need to correct for multiple testing. DESeq2 uses the Benjamini-Hochberg False Discovery Rate (FDR) correction, and these values are reported in the column <strong><code>padj</code></strong> (i.e., adjusted p-value).</p></li>
</ul>
<p>A summary of this information about each column can be seen by running the <code>mcols()</code> function:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mcols</span>(res)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="na-values-in-the-results-table" class="level3" data-number="6.2">
<h3 data-number="6.2" class="anchored" data-anchor-id="na-values-in-the-results-table"><span class="header-section-number">6.2</span> <code>NA</code> values in the results table</h3>
<p>Some values in the results table can be set to <code>NA</code> for one of the following reasons:</p>
<ul>
<li><p>If a gene contains a sample with a count <strong>outlier</strong>, both the p-value and adjusted p-value will be set to <code>NA</code>. (DESeq2 performs outlier detection using Cook’s distance.)</p></li>
<li><p>If all samples have <strong>zero counts</strong> for a given gene, the <code>baseMean</code> column will be zero, and the log2-fold change estimates, p-value and adjusted p-value will all be set to <code>NA</code>.</p></li>
<li><p>DESeq2 also automatically filters genes with a <strong>low mean count</strong> in the sense that it does not include them in the multiple testing correction. Therefore, in such cases, the p-value will not be <code>NA</code>, but the <em>adjusted</em> p-value will be.</p>
<p>Because we have very low power to detect differential expression for such low-count genes, it is beneficial to remove them prior to the multiple testing correction: that way, the correction becomes less severe for the remaining genes.</p></li>
</ul>
<p>Let’s see how many genes have <code>NA</code> p-values:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Number of genes with NA p-value:</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(<span class="fu">is.na</span>(res<span class="sc">$</span>pvalue))</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a><span class="co"># As a proportion of the total number of genes in the test:</span></span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(<span class="fu">is.na</span>(res<span class="sc">$</span>pvalue)) <span class="sc">/</span> <span class="fu">nrow</span>(res)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>And <code>NA</code> adjusted p-values:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Number of genes with NA p-value:</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(<span class="fu">is.na</span>(res<span class="sc">$</span>padj))</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a><span class="co"># As a proportion of the total number of genes in the test:</span></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(<span class="fu">is.na</span>(res<span class="sc">$</span>padj)) <span class="sc">/</span> <span class="fu">nrow</span>(res)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
</section>
</section>
<section id="de-analysis-contrast-two-custom-groups" class="level2" data-number="7">
<h2 data-number="7" class="anchored" data-anchor-id="de-analysis-contrast-two-custom-groups"><span class="header-section-number">7</span> DE analysis – contrast two custom groups</h2>
<p>Using the <code>resultsNames</code> function, we can see which pairwise contrasts between different levels of the factor are available (though it is not displayed in a particularly readable fashion):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">resultsNames</span>(dds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Not all pairwise contrasts between the 5 levels in our <code>group</code> factor are available here: instead, <code>control_mock</code>, which we set as the reference level, is being compared with the other 3 levels. (However, we can make other pairwise comparisons, too.)</p>
<p>Above, we looked at the results for the <em>last</em> of these comparisons (<code>group_Ri_myb_vs_control_mock</code>, i.e.&nbsp;“Ri_myb” vs.&nbsp;“control_mock”), simply because DESeq2 will show the last comparison by default when calling the <code>results()</code> function.</p>
<p>To see the results table for one of the other 3 comparisons, we pass a vector to the <code>contrast</code> argument of the <code>results()</code> function with the factor (<code>group</code>) and the two levels to be contrasted. For example, to see the results for “Ri_mock” vs.&nbsp;“control_mock”:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Here, we could specify *any* pairwise contrast,</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="co"># not just the ones with "control_mock" that resultsNames() prints as seen above.</span></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>my_contrast <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Ri_mock"</span>, <span class="st">"control_mock"</span>)</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">results</span>(dds,</span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>               <span class="at">contrast =</span> <span class="fu">c</span>(<span class="st">"group"</span>, my_contrast))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>How many adjusted p-values were less than 0.1?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(res<span class="sc">$</span>padj <span class="sc">&lt;</span> <span class="fl">0.1</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
</section>
<section id="visually-exploring-the-results" class="level2" data-number="8">
<h2 data-number="8" class="anchored" data-anchor-id="visually-exploring-the-results"><span class="header-section-number">8</span> Visually exploring the results</h2>
<p>We will create a few plots, by way of example, of the results for the “Ri_mock” versus “control_mock” comparison, which we extracted above.</p>
<section id="volcano-plot" class="level3" data-number="8.1">
<h3 data-number="8.1" class="anchored" data-anchor-id="volcano-plot"><span class="header-section-number">8.1</span> Volcano plot</h3>
<p>For a nice overview of the results, we can plot a so-called “volcano plot”.</p>
</section>
<section id="plot-specific-genes" class="level3" data-number="8.2">
<h3 data-number="8.2" class="anchored" data-anchor-id="plot-specific-genes"><span class="header-section-number">8.2</span> Plot specific genes</h3>
<p>We can also create plot of expression levels for individual genes. That is especially interesting for genes with highly significant differential expression.</p>
<p>Let’s plot the top-5 most significantly differentially expressed genes:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># First, we select the 5 genes with the lowest adjusted p-value:</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>top5 <span class="ot">&lt;-</span> <span class="fu">row.names</span>(res[<span class="fu">order</span>(res<span class="sc">$</span>padj)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>], ])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Then we create a function to make a plot for a single gene:</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>plotgene <span class="ot">&lt;-</span> <span class="cf">function</span>(geneID, dds) {</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>  d <span class="ot">&lt;-</span> <span class="fu">plotCounts</span>(dds,</span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>                  <span class="at">gene =</span> geneID,</span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>                  <span class="at">intgroup =</span> <span class="st">"group"</span>,</span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a>                  <span class="at">returnData =</span> <span class="cn">TRUE</span>)</span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a>  p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(d, <span class="fu">aes</span>(<span class="at">x =</span> group, <span class="at">y =</span> count)) <span class="sc">+</span></span>
<span id="cb33-10"><a href="#cb33-10" aria-hidden="true" tabindex="-1"></a>          <span class="fu">geom_point</span>(<span class="at">position =</span> <span class="fu">position_jitter</span>(<span class="at">w =</span> <span class="fl">0.1</span>, <span class="at">h =</span> <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb33-11"><a href="#cb33-11" aria-hidden="true" tabindex="-1"></a>          <span class="fu">labs</span>(<span class="at">title =</span> geneID) <span class="sc">+</span></span>
<span id="cb33-12"><a href="#cb33-12" aria-hidden="true" tabindex="-1"></a>          <span class="fu">theme_bw</span>()</span>
<span id="cb33-13"><a href="#cb33-13" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb33-14"><a href="#cb33-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(p)</span>
<span id="cb33-15"><a href="#cb33-15" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Finally, we use <code>sapply()</code> to apply this function to each of our genes in the <code>top5</code> vector.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>none <span class="ot">&lt;-</span> <span class="fu">sapply</span>(top5, plotgene, dds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>If we wanted to, we could easily create plots for 100s of genes, this way.</p>
</section>
<section id="heatmap" class="level3" data-number="8.3">
<h3 data-number="8.3" class="anchored" data-anchor-id="heatmap"><span class="header-section-number">8.3</span> Heatmap</h3>
<p>We can create heatmaps with the <code>pheatmap</code> function. Let’s start by creating a function that will plot a heatmap given a vector of gene IDs and a DESeq2 object <code>dds</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>plot_heatmap <span class="ot">&lt;-</span> <span class="cf">function</span>(geneIDs, dds) {</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>  ntd <span class="ot">&lt;-</span> <span class="fu">assay</span>(<span class="fu">normTransform</span>(dds))</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>  ntd_sel <span class="ot">&lt;-</span> ntd[<span class="fu">match</span>(geneIDs, <span class="fu">rownames</span>(ntd)), ]</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>  df_meta <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">colData</span>(dds)[, <span class="fu">c</span>(<span class="st">"AMF"</span>, <span class="st">"Treatment"</span>)])</span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pheatmap</span>(ntd_sel,</span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a>         <span class="at">cluster_rows =</span> <span class="cn">FALSE</span>,</span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a>         <span class="at">cluster_cols =</span> <span class="cn">FALSE</span>,</span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a>         <span class="at">show_rownames =</span> <span class="cn">FALSE</span>,</span>
<span id="cb35-12"><a href="#cb35-12" aria-hidden="true" tabindex="-1"></a>         <span class="at">annotation_col =</span> df_meta)</span>
<span id="cb35-13"><a href="#cb35-13" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now, we can easily create a heatmap for the top-20 most highly differentially expressed genes:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>top20_DE <span class="ot">&lt;-</span> <span class="fu">row.names</span>(res[<span class="fu">order</span>(res<span class="sc">$</span>padj)[<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>], ])</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_heatmap</span>(top20_DE, dds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Or for the 20 most highly expressed genes:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>top20_hi_idx <span class="ot">&lt;-</span> <span class="fu">order</span>(<span class="fu">rowMeans</span>(<span class="fu">counts</span>(dds, <span class="at">normalized =</span> <span class="cn">TRUE</span>)),</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>                  <span class="at">decreasing =</span> <span class="cn">TRUE</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>]</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>top20_hi <span class="ot">&lt;-</span> <span class="fu">row.names</span>(dds)[top20_hi_idx]</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_heatmap</span>(top20_hi, dds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="export-the-results" class="level3" data-number="8.4">
<h3 data-number="8.4" class="anchored" data-anchor-id="export-the-results"><span class="header-section-number">8.4</span> Export the results</h3>
<p>Let’s save the results dataframe to file.</p>
<p>Note that it will only contain the results for one comparison. Also, if we write the results dataframe to file, we won’t be able to tell from the file what the comparison is, so let’s store that in two columns:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>my_contrast</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>my_contrast_pasted <span class="ot">&lt;-</span> <span class="fu">paste0</span>(my_contrast, <span class="at">collapse =</span> <span class="st">"_vs_"</span>)</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>my_contrast_pasted</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>res<span class="sc">$</span>level1 <span class="ot">&lt;-</span> my_contrast[<span class="dv">1</span>]</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>res<span class="sc">$</span>level2 <span class="ot">&lt;-</span> my_contrast[<span class="dv">2</span>]</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>(<span class="fu">head</span>(res))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we can write <code>res</code> to file:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>res_file <span class="ot">&lt;-</span> <span class="fu">file.path</span>(outdir, <span class="fu">paste0</span>(my_contrast_pasted, <span class="st">'_all-res.txt'</span>))</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(res, res_file,</span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>            <span class="at">sep =</span> <span class="st">'</span><span class="sc">\t</span><span class="st">'</span>, <span class="at">row.names =</span> <span class="cn">TRUE</span>, <span class="at">quote =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
</section>
</section>
<section id="de-analysis-with-two-factors" class="level2" data-number="9">
<h2 data-number="9" class="anchored" data-anchor-id="de-analysis-with-two-factors"><span class="header-section-number">9</span> DE analysis – with two factors</h2>
<section id="controlling-for-one-factor" class="level3" data-number="9.1">
<h3 data-number="9.1" class="anchored" data-anchor-id="controlling-for-one-factor"><span class="header-section-number">9.1</span> Controlling for one factor</h3>
<p>Say we wanted to analyze the effect of “mock” versus “myb” (“Treatment” column) while controlling for the effects of “control” versus “Ri” (“AMF” column).</p>
<p>Let’s start by turning “Treatment” and “AMF” into factors, and saving a new DESeq2 object:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert Treatment and AMF into factors: </span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>dds_raw<span class="sc">$</span>Treatment <span class="ot">&lt;-</span> <span class="fu">relevel</span>(<span class="fu">factor</span>(dds_raw<span class="sc">$</span>Treatment), <span class="at">ref =</span> <span class="st">"mock"</span>)</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>dds_raw<span class="sc">$</span>AMF <span class="ot">&lt;-</span> <span class="fu">relevel</span>(<span class="fu">factor</span>(dds_raw<span class="sc">$</span>AMF), <span class="at">ref =</span> <span class="st">"control"</span>)</span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Save a new object:</span></span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a>dds_2f_raw <span class="ot">&lt;-</span> dds_raw</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To include both factors, we use a <code>+</code> in the formula. Note that the order matters: using <code>AMF + Treatment</code>, we test for the effect of “Treatment” (the last factor), while controlling for the effect of AMF (the first factor).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="fu">design</span>(dds_2f_raw) <span class="ot">&lt;-</span> <span class="fu">formula</span>(<span class="sc">~</span> AMF <span class="sc">+</span> Treatment)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Run DESeq with the new design:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>dds_2f <span class="ot">&lt;-</span> <span class="fu">DESeq</span>(dds_2f_raw)</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">results</span>(dds_2f)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>How many adjusted p-values were less than 0.1?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(res<span class="sc">$</span>padj <span class="sc">&lt;</span> <span class="fl">0.1</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>


</section>
</section>

<a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>