<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Jelmer Poelstra">
<meta name="dcterms.date" content="2024-01-26">

<title>ENTMLGY 6703 - Jelmer’s lectures - Week 3 lab — part 1:Computational Infrastructure</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<link href="../site_libs/quarto-contrib/fontawesome6-0.1.0/all.css" rel="stylesheet">
<link href="../site_libs/quarto-contrib/fontawesome6-0.1.0/latex-fontsize.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">ENTMLGY 6703 - Jelmer’s lectures</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-week-3-hts" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Week 3: HTS</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-week-3-hts">    
        <li>
    <a class="dropdown-item" href="../lectures/sequencing.html" rel="" target="">
 <span class="dropdown-text">Lecture: High-throughput Sequencing &amp; Genomes</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../labs/FASTQ.html" rel="" target="">
 <span class="dropdown-text">Lab: Genome and HTS data</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-week-4-rna-seq" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Week 4: RNA-seq</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-week-4-rna-seq">    
        <li>
    <a class="dropdown-item" href="../lectures/rnaseq.html" rel="" target="">
 <span class="dropdown-text">Lecture: RNA-seq</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../labs/DE.html" rel="" target="">
 <span class="dropdown-text">Lab: Differential expression analysis</span></a>
  </li>  
    </ul>
  </li>
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../about.html" rel="" target="">
 <span class="menu-text">About</span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label=""><i class="bi bi-github"></i></a>
      <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="quarto-navigation-tool-dropdown-0">
          <li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://github.com/osu-codeclub/osu-codeclub.github.io">
            Source Code
            </a>
          </li>
          <li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://github.com/osu-codeclub/osu-codeclub.github.io/issues/new">
            Report an Issue
            </a>
          </li>
      </ul>
    </div>
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction-to-the-lab" id="toc-introduction-to-the-lab" class="nav-link active" data-scroll-target="#introduction-to-the-lab"><span class="header-section-number">1</span> Introduction to the lab</a></li>
  <li><a href="#computational-infrastructure-overview" id="toc-computational-infrastructure-overview" class="nav-link" data-scroll-target="#computational-infrastructure-overview"><span class="header-section-number">2</span> Computational infrastructure overview</a></li>
  <li><a href="#the-ohio-supercomputer-center-osc" id="toc-the-ohio-supercomputer-center-osc" class="nav-link" data-scroll-target="#the-ohio-supercomputer-center-osc"><span class="header-section-number">3</span> The Ohio Supercomputer Center (OSC)</a>
  <ul class="collapse">
  <li><a href="#introduction-to-osc" id="toc-introduction-to-osc" class="nav-link" data-scroll-target="#introduction-to-osc"><span class="header-section-number">3.1</span> Introduction to OSC</a></li>
  <li><a href="#the-osc-ondemand-web-portal" id="toc-the-osc-ondemand-web-portal" class="nav-link" data-scroll-target="#the-osc-ondemand-web-portal"><span class="header-section-number">3.2</span> The OSC OnDemand web portal</a></li>
  </ul></li>
  <li><a href="#the-vs-code-text-editor" id="toc-the-vs-code-text-editor" class="nav-link" data-scroll-target="#the-vs-code-text-editor"><span class="header-section-number">4</span> The VS Code text editor</a>
  <ul class="collapse">
  <li><a href="#what-is-vs-code" id="toc-what-is-vs-code" class="nav-link" data-scroll-target="#what-is-vs-code"><span class="header-section-number">4.1</span> What is VS Code?</a></li>
  <li><a href="#connecting-to-vs-code" id="toc-connecting-to-vs-code" class="nav-link" data-scroll-target="#connecting-to-vs-code"><span class="header-section-number">4.2</span> Connecting to VS Code</a></li>
  <li><a href="#the-vs-code-user-interface" id="toc-the-vs-code-user-interface" class="nav-link" data-scroll-target="#the-vs-code-user-interface"><span class="header-section-number">4.3</span> The VS Code User Interface</a></li>
  </ul></li>
  <li><a href="#the-unix-shell" id="toc-the-unix-shell" class="nav-link" data-scroll-target="#the-unix-shell"><span class="header-section-number">5</span> The Unix shell</a>
  <ul class="collapse">
  <li><a href="#what-is-the-unix-shell" id="toc-what-is-the-unix-shell" class="nav-link" data-scroll-target="#what-is-the-unix-shell"><span class="header-section-number">5.1</span> What is the Unix shell?</a></li>
  <li><a href="#first-steps-in-the-unix-shell" id="toc-first-steps-in-the-unix-shell" class="nav-link" data-scroll-target="#first-steps-in-the-unix-shell"><span class="header-section-number">5.2</span> First steps in the Unix shell</a></li>
  <li><a href="#cd-and-command-actions-defaults-and-arguments" id="toc-cd-and-command-actions-defaults-and-arguments" class="nav-link" data-scroll-target="#cd-and-command-actions-defaults-and-arguments"><span class="header-section-number">5.3</span> <code>cd</code> and command actions, defaults, and arguments</a></li>
  <li><a href="#ls-and-command-options" id="toc-ls-and-command-options" class="nav-link" data-scroll-target="#ls-and-command-options"><span class="header-section-number">5.4</span> <code>ls</code> and command <em>options</em></a></li>
  <li><a href="#unix-shell-recap" id="toc-unix-shell-recap" class="nav-link" data-scroll-target="#unix-shell-recap"><span class="header-section-number">5.5</span> Unix shell recap</a></li>
  </ul></li>
  <li><a href="#fastq-files" id="toc-fastq-files" class="nav-link" data-scroll-target="#fastq-files"><span class="header-section-number">6</span> FASTQ files</a></li>
  <li><a href="#running-fastqc" id="toc-running-fastqc" class="nav-link" data-scroll-target="#running-fastqc"><span class="header-section-number">7</span> Running FastQC</a>
  <ul class="collapse">
  <li><a href="#what-is-fastqc" id="toc-what-is-fastqc" class="nav-link" data-scroll-target="#what-is-fastqc"><span class="header-section-number">7.1</span> What is FastQC?</a></li>
  <li><a href="#running-fastqc-1" id="toc-running-fastqc-1" class="nav-link" data-scroll-target="#running-fastqc-1"><span class="header-section-number">7.2</span> Running FastQC</a></li>
  </ul></li>
  <li><a href="#interpreting-fastqcs-output" id="toc-interpreting-fastqcs-output" class="nav-link" data-scroll-target="#interpreting-fastqcs-output"><span class="header-section-number">8</span> Interpreting FastQC’s output</a></li>
  <li><a href="#next-steps" id="toc-next-steps" class="nav-link" data-scroll-target="#next-steps"><span class="header-section-number">9</span> Next steps</a>
  <ul class="collapse">
  <li><a href="#next-steps-in-an-rna-seq-workflow" id="toc-next-steps-in-an-rna-seq-workflow" class="nav-link" data-scroll-target="#next-steps-in-an-rna-seq-workflow"><span class="header-section-number">9.1</span> Next steps in an RNA-seq workflow</a></li>
  <li><a href="#next-steps-to-run-the-analyses-more-efficiently" id="toc-next-steps-to-run-the-analyses-more-efficiently" class="nav-link" data-scroll-target="#next-steps-to-run-the-analyses-more-efficiently"><span class="header-section-number">9.2</span> Next steps to run the analyses more efficiently</a></li>
  </ul></li>
  <li><a href="#to-learn-more" id="toc-to-learn-more" class="nav-link" data-scroll-target="#to-learn-more"><span class="header-section-number">10</span> To learn more</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Week 3 lab — part 1:<br>Computational Infrastructure</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Jelmer Poelstra </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">January 26, 2024</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<p><br></p>
<section id="introduction-to-the-lab" class="level2" data-number="1">
<h2 data-number="1" class="anchored" data-anchor-id="introduction-to-the-lab"><span class="header-section-number">1</span> Introduction to the lab</h2>
<p>x</p>
<p><br></p>
</section>
<section id="computational-infrastructure-overview" class="level2" data-number="2">
<h2 data-number="2" class="anchored" data-anchor-id="computational-infrastructure-overview"><span class="header-section-number">2</span> Computational infrastructure overview</h2>
<p>Due in large part to the amount of data involved, you don’t typically use a laptop or desktop computer to work with HTS data, or with genomics and transcriptomics data more generally.</p>
<p>Additionally, most of the specialized tools (i.e., programs / software) that help you analyze your data can only be run through a “command-line interface” (CLI).</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Additional advantages
</div>
</div>
<div class="callout-body-container callout-body">
<p>And even those that have a “graphical user interface” (GUI) are more efficiently and reproducibly run through a CLI:</p>
<ul>
<li><p><strong>Efficiency</strong> — A CLI allows you to write a simple loop to run it in the same way for many samples. (In combination with the computing power of a supercomputer, this in turn allows you to process those hundreds of samples <em>in parallel</em>.)</p></li>
<li><p><strong>Reproducibility</strong> — You can easily save all commands and scripts which would allow you to rerun a project rather straightforwardly.</p></li>
</ul>
</div>
</div>
<p>Therefore, a typical computational infrastructure to do what we may call “command-line genomics” involves:</p>
<ol type="1">
<li>A <strong>supercomputer</strong><a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a> — in our case, the Ohio Supercomputer Center (OSC)</li>
<li>A <strong>text editor</strong> — I recommend and will demonstrate <em>VS Code</em></li>
<li>The <strong>Unix shell</strong> (terminal)</li>
<li><strong>R</strong> (or perhaps <em>Python</em>) for interactive statistical analysis and visualization.</li>
</ol>
<p>Below, we will go through the first three of these. In the lab next week, we will cover the fourth.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>[Often dividable in 2 sections: compute-heavy command-line programs, then not compute-heavy with R.]</p>
<p>A genomics project usually involves sequentially running an array of bioinformatics programs (or “tools”). For instance, as we’ll discuss next week, an RNA-seq project may include:</p>
<blockquote class="blockquote">
<p>raw read QC =&gt; raw read trimming =&gt; read alignment mapping =&gt; gene counting</p>
</blockquote>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Can I <em>avoid</em> all of this?
</div>
</div>
<div class="callout-body-container callout-body">
<p>If you will often do genomics projects like the ones mentioned above, it’s hard to avoid using the infrastructure as described. But here are some conditions in which you might reasonably avoid it:</p>
<ul>
<li>You’re doing a single genomics project, your main research focus is elsewhere</li>
<li>You have data which can be analyzed with a relatively small command-line-based part, such as metabarcoding or RNA-seq.</li>
</ul>
<p>In such cases, you might be able to get someone else to do the command-line work, or you could try <strong><em>Galaxy</em></strong>, a cloud-based bioinformatics platform with a web browser interface and no coding.</p>
</div>
</div>
<p><br></p>
</section>
<section id="the-ohio-supercomputer-center-osc" class="level2" data-number="3">
<h2 data-number="3" class="anchored" data-anchor-id="the-ohio-supercomputer-center-osc"><span class="header-section-number">3</span> The Ohio Supercomputer Center (OSC)</h2>
<section id="introduction-to-osc" class="level3" data-number="3.1">
<h3 data-number="3.1" class="anchored" data-anchor-id="introduction-to-osc"><span class="header-section-number">3.1</span> Introduction to OSC</h3>
<section id="what-is-a-supercomputer" class="level4">
<h4 class="anchored" data-anchor-id="what-is-a-supercomputer">What is a supercomputer?</h4>
<p>A supercomputer is a highly interconnected set of many computer processors and storage units. You can think of it simply as a <strong>network of computers</strong>.</p>
<p>Supercomputers are also commonly referred to as <em>High-Performance Computing (<strong>HPC</strong>) clusters</em> or simply <strong>compute clusters</strong>. This is what Owens, one of the OSC supercomputers, physically looks like:</p>
<p align="center">
<img src="img_hts/owens.jpg" width="50%">
</p>
</section>
<section id="why-do-i-need-a-supercomputer" class="level4">
<h4 class="anchored" data-anchor-id="why-do-i-need-a-supercomputer">Why do I need a supercomputer?</h4>
<ul>
<li>Often, your genomics <strong>dataset is too large</strong> to be handled efficiently, or even at all, by a laptop or desktop computer.</li>
<li>To <strong>speed up</strong> <em>long-running analyses</em> by using more computing power, and <em>repeated analyses</em> (like the independent alignment of reads for different samples) by running them in parallel.</li>
<li>It’s also a great place to <strong>store</strong> large amounts of data.</li>
</ul>
</section>
<section id="the-ohio-supercomputer-center-osc-1" class="level4">
<h4 class="anchored" data-anchor-id="the-ohio-supercomputer-center-osc-1">The Ohio Supercomputer Center (OSC)</h4>
<p align="center">
<img src="img_hts/osc_logo.png" width="70%">
</p>
<p>The <strong>Ohio Supercomputer Center (OSC)</strong> provides computing resources to researchers (and others) across Ohio.</p>
<p>OSC has two supercomputers/clusters (named Owens and Pitzer), and lots of infrastructure for their usage. Research usage is charged but at <a href="https://www.osc.edu/content/academic_fee_model_faq">heavily subsidized rates</a>, and most or all OSU colleges absorb these costs at the college level (!).</p>
</section>
<section id="the-structure-of-a-supercomputer" class="level4">
<h4 class="anchored" data-anchor-id="the-structure-of-a-supercomputer">The Structure of a Supercomputer</h4>
<p>We can think of a supercomputer as having three main parts:</p>
<ul>
<li><strong>File Systems</strong>: Where files are stored (these are shared between the two OSC clusters)</li>
<li><strong>Login Nodes</strong>: The handful of computers everyone shares after logging in</li>
<li><strong>Compute Nodes</strong>: The many computers you can reserve to run your analyses</li>
</ul>
<p align="center">
<img src="img_hts/cluster_overview_ed.png" width="85%">
</p>
</section>
<section id="what-is-different-on-a-supercomputer-like-at-osc" class="level4">
<h4 class="anchored" data-anchor-id="what-is-different-on-a-supercomputer-like-at-osc">What is different on a supercomputer like at OSC?</h4>
<p>Compared to command-line computing on a laptop or desktop, the following aspects are different when working on a supercomputer like at OSC?</p>
<ul>
<li><p><strong>Operating system</strong></p></li>
<li><p><strong>Login versus compute nodes</strong> (node ≈ computer)<br>
“Login nodes”, the nodes you end up on after logging in, are not meant for heavy computing and you have to <em>request access to “compute nodes”</em> to run most analyses.</p></li>
<li><p><strong>Software</strong><br>
Because you don’t have administrator rights, and because the system is shared by so many people, you generally can’t install and use software “the regular way”.</p>
<ul>
<li>For system-wide installed software: some programs are directly available, while more specialized programs like for bioinformatics <em>need to be explicitly loaded</em>.</li>
<li>If something is not installed, it’s best to ask OSC, or use solutions like <em>Conda</em> or containers that are outside the scope of this introduction.</li>
</ul></li>
<li><p><strong>“Non-interactive” usage is common, using a job scheduler (Slurm)</strong><br>
You submit your scripts to the Slurm queue and monitor the resulting jobs. We will …</p></li>
</ul>
<p><br></p>
</section>
</section>
<section id="the-osc-ondemand-web-portal" class="level3" data-number="3.2">
<h3 data-number="3.2" class="anchored" data-anchor-id="the-osc-ondemand-web-portal"><span class="header-section-number">3.2</span> The OSC OnDemand web portal</h3>
<p>The OSC OnDemand web portal allows you to use a web browser to access OSC resources such as:</p>
<ul>
<li>A file browser where you can also create and rename folders and files, etc.</li>
<li>A Unix shell</li>
<li>A host of “Interactive Apps”: programs such as RStudio, Jupyter, VS Code and QGIS.</li>
</ul>
<p>When you go to <a href="https://ondemand.osc.edu" class="uri">https://ondemand.osc.edu</a>, you first need to <strong>log in</strong> with your OSC (not OSU!) credentials. After that, you should see a landing page similar to the one below:</p>
<p align="center">
<img src="img_hts/ondemand_home.png" width="95%">
</p>
<p>The main part of the page (below the logo) only contains some general OSC messages and updates — what we will focus on instead are some of the options in the <strong>blue bar along the top</strong>.</p>
<section id="file-system-access" class="level4">
<h4 class="anchored" data-anchor-id="file-system-access">File System Access</h4>
<p>Let’s start with <strong>Files</strong>. Hovering over this dropdown menu gives a list of directories you have access to. If your account is brand new, you should only have three:</p>
<ul>
<li>A Home directory (starts with <code>/users</code>)</li>
</ul>
<p>Two directories directly associated with the OSC “project” <code>PAS2250</code> that I added you to: <a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a></p>
<ul>
<li>A “project” directory (starts with <code>/fs/ess</code>) — permanent, backed-up storage</li>
<li>A “scratch” directory (starts with <code>/fs/scratch</code>) — temporary storage</li>
</ul>
<p>I’ll select the scratch directory, <code>/fs/scratch/PAS2250</code>:</p>
<p align="center">
<img src="img_hts/ondemand_files_PAS2250.png" width="30%">
</p>
<p>Once there, I can see a list of directories and files inside this Project directory, and I can click on the directories to explore the contents further:</p>
<p>[IMAGE TBA]</p>
<p>This interface is <strong>much like the file browser on your own computer</strong>, so you can also create, delete, move and copy files and folders, and even upload (from your computer to OSC) and download (from OSC your computer) files<a href="#fn3" class="footnote-ref" id="fnref3" role="doc-noteref"><sup>3</sup></a> — see the buttons across the top.</p>
</section>
<section id="on-your-own-create-your-own-folder" class="level4 exercise">
<h4 class="anchored" data-anchor-id="on-your-own-create-your-own-folder">On Your Own: Create your own folder</h4>
<p>TBA</p>
</section>
<section id="unix-shell-access-under-clusters" class="level4">
<h4 class="anchored" data-anchor-id="unix-shell-access-under-clusters">Unix Shell Access (under Clusters)</h4>
<p>Interacting with a supercomputer in a point-and-click manner only goes so far. Using a supercomputer effectively requires interacting with the system using a command-line interface (<strong>CLI</strong>) of a Unix shell.</p>
<p>Under the <strong>Clusters</strong> dropdown menu, you can access a Unix shell either on Owens or Pitzer:</p>
<p align="center">
<img src="img_hts/ondemand_shell_select.png" width="50%">
</p>
<p>I’m selecting a shell on the Pitzer supercomputer, which will open a new browser tab looking like this:</p>
<p align="center">
<img src="img_hts/ondemand_shell2.png" width="95%">
</p>
<p>We most commonly interact with a supercomputer using a Unix shell, and we’ll learn about the basics of doing so soon. However, we’ll mostly be accessing a Unix shell in a different manner, namely <strong>inside the VS Code text editor</strong>, which also gives us some additional functionality in a user-friendly way.</p>
</section>
<section id="interactive-apps" class="level4">
<h4 class="anchored" data-anchor-id="interactive-apps">Interactive Apps</h4>
<p>We can access programs with Graphical User Interfaces (GUIs; point-and-click interfaces) via the <strong>Interactive Apps</strong> dropdown menu <a href="#fn4" class="footnote-ref" id="fnref4" role="doc-noteref"><sup>4</sup></a> — let’s select VS Code using the “Code Server” button:</p>
<p align="center">
<img src="img_hts/ondemand_vscode_select.png" width="32%">
</p>
<p><br></p>
</section>
</section>
</section>
<section id="the-vs-code-text-editor" class="level2" data-number="4">
<h2 data-number="4" class="anchored" data-anchor-id="the-vs-code-text-editor"><span class="header-section-number">4</span> The VS Code text editor</h2>
<section id="what-is-vs-code" class="level3" data-number="4.1">
<h3 data-number="4.1" class="anchored" data-anchor-id="what-is-vs-code"><span class="header-section-number">4.1</span> What is VS Code?</h3>
<p>VS Code (in full, Visual Studio Code) is basically a fancy text editor.</p>
<p>To emphasize the additional functionality relative to basic text editors like Notepad and TextEdit, editors like VS Code are also referred to as “<strong>IDEs</strong>”: <em>Integrated Development Environments</em>. The RStudio program is another good example of an IDE. For our purposes:</p>
<ul>
<li>VS code will be our IDE for Unix shell code (this week)</li>
<li>RStudio will be our IDE for R (in the differential expression lab next week)</li>
</ul>
<p>Conveniently, we can use a version of this editor (sometimes referred to as <em>Code Server</em>, like in the Interactive Apps menu) in our browser via the OSC OnDemand website.</p>
</section>
<section id="connecting-to-vs-code" class="level3" data-number="4.2">
<h3 data-number="4.2" class="anchored" data-anchor-id="connecting-to-vs-code"><span class="header-section-number">4.2</span> Connecting to VS Code</h3>
<p>Because “Interactive Apps” like VS Code and RStudio <strong>run on compute nodes</strong>, we need to fill out a form and specify the following details (as shown in the image below) to reserve such a node:</p>
<ul>
<li>The <strong><em>OSC Project</em></strong> that should be billed for the compute resource usage</li>
<li>The <strong><em>amount of time</em></strong> we want to make a reservation for <a href="#fn5" class="footnote-ref" id="fnref5" role="doc-noteref"><sup>5</sup></a></li>
<li>The <strong>“<em>working directory</em>”</strong> (starting location in the file system) for the program</li>
<li>The <strong>version</strong> of VS Code</li>
</ul>
<p>[TBA - form with correct values filled out]</p>
<p>Click on <strong>Launch</strong> at the bottom and this will send your request to the compute job scheduler. First, your job will be “Queued” — that is, waiting for the job scheduler to allocate resources on the compute nodes to it:</p>
<p align="center">
<img src="img_hts/ondemand_vscode_queued.png" width="80%">
</p>
<p>In general, it should be granted resources within a few seconds (the card will then say “Starting”), and be ready for usage (“Running”) in another couple of seconds:</p>
<p align="center">
<img src="img_hts/ondemand_vscode_running.png" width="80%">
</p>
<p>Then, you can click on the blue <strong>Connect to VS Code</strong> button to open a new browser tab that runs VS Code. When VS Code opens, you may get these two pop-ups — click “Yes” and “Don’t Show Again”, respectively:</p>
<div class="columns">
<div class="column" style="width:55%;">
<p align="center">
<img src="img_HTS/vscode-trust.png" width="90%">
</p>
</div><div class="column" style="width:45%;">
<p align="center">
<img src="img_HTS/vscode-git.png" width="90%">
</p>
</div>
</div>
</section>
<section id="the-vs-code-user-interface" class="level3" data-number="4.3">
<h3 data-number="4.3" class="anchored" data-anchor-id="the-vs-code-user-interface"><span class="header-section-number">4.3</span> The VS Code User Interface</h3>
<p align="left">
<img src="img_HTS/vscode-welcome_ed.png" width="80%">
</p>
<section id="side-bars" class="level4">
<h4 class="anchored" data-anchor-id="side-bars">Side bars</h4>
<p>The <strong>Activity Bar</strong> (narrow side bar) on the far left has:</p>
<ul>
<li>A &nbsp; <i class="fa-solid fa-bars" aria-label="bars"></i> &nbsp; (“hamburger menu”), which has menu items like <code>File</code> that you often find in a top bar.</li>
<li>A &nbsp; <i class="fa-solid fa-cog" aria-label="cog"></i> &nbsp; (cog wheel icon) in the bottom, through which you can mainly access <em>settings</em>.</li>
<li>Icons that serve to switch between different options for the <strong>(wide) Side Bar</strong> — we’ll only use:
<ul>
<li><em>Explorer</em>: File browser (and, e.g., an outline for the active file)</li>
<li><em>Extensions</em>: To install extensions (we’ll install one later)</li>
</ul></li>
</ul>
</section>
<section id="editor-pane-and-welcome-document" class="level4">
<h4 class="anchored" data-anchor-id="editor-pane-and-welcome-document">Editor pane and <code>Welcome</code> document</h4>
<p>The main part of the VS Code is the <em>editor pane</em>. Whenever you open VS Code, a tab with a <code>Welcome</code> document is automatically opened. This provides some help and e.g.&nbsp;a handy overview of recently opened folders.</p>
<p>We can also use the <code>Welcome</code> document <a href="#fn6" class="footnote-ref" id="fnref6" role="doc-noteref"><sup>6</sup></a> to open a new text file by clicking <code>New file</code> below <code>Start</code> (alternatively, click &nbsp; <i class="fa-solid fa-bars" aria-label="bars"></i> &nbsp; =&gt; &nbsp; <code>File</code> &nbsp; =&gt; &nbsp; <code>New File</code>), which opens as a second “tab” in the editor pane.</p>
</section>
<section id="terminal-with-a-unix-shell" class="level4">
<h4 class="anchored" data-anchor-id="terminal-with-a-unix-shell">Terminal (with a Unix shell)</h4>
<p><i class="fa-solid fa-user-edit" aria-label="user-edit"></i> <strong>Open a terminal</strong> by clicking &nbsp; <i class="fa-solid fa-bars" aria-label="bars"></i> &nbsp; =&gt; <code>Terminal</code> =&gt; <code>New Terminal</code>.</p>
</section>
<section id="on-your-own-try-a-few-color-themes" class="level4 unnumbered exercise">
<h4 class="unnumbered anchored" data-anchor-id="on-your-own-try-a-few-color-themes">On Your Own: Try a few color themes</h4>
<p>Open the Command Palette and start typing “color theme”, and you’ll see the relevant option pop up.</p>
<p>Then, try out a few themes and see what you like!</p>
<p>(You can also access the Color Themes option via &nbsp; <i class="fa fa-cog"></i> &nbsp; =&gt; <code>Color Theme</code>.)</p>
</section>
<p><br></p>
</section>
</section>
<section id="the-unix-shell" class="level2" data-number="5">
<h2 data-number="5" class="anchored" data-anchor-id="the-unix-shell"><span class="header-section-number">5</span> The Unix shell</h2>
<section id="what-is-the-unix-shell" class="level3" data-number="5.1">
<h3 data-number="5.1" class="anchored" data-anchor-id="what-is-the-unix-shell"><span class="header-section-number">5.1</span> What is the Unix shell?</h3>
<p>A computer’s <strong>shell</strong> is the interface in a Terminal window that allows you to interact with your computer by <em>typing commands</em> rather than pointing-and-clicking.</p>
<p>It is also referred to as the the “<strong>command line</strong>” — with “command-line tools/programs” being software that is run using shell commands.</p>
<p>Finally, the <strong>Unix shell</strong>, then, is the shell of Unix-based computers, which include Mac and Linux (but not Windows) operating systems. <a href="#fn7" class="footnote-ref" id="fnref7" role="doc-noteref"><sup>7</sup></a></p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Many bioinformatics programs are basically specialized commands
</div>
</div>
<div class="callout-body-container callout-body">
<p>In many ways, as mentioned in the box above, you can think of using a command-line bioinformatics program as using just another command.</p>
<p>Therefore, our general skills with Unix commands will very much extend to using command-line bioinformatics tools!</p>
</div>
</div>
</section>
<section id="first-steps-in-the-unix-shell" class="level3" data-number="5.2">
<h3 data-number="5.2" class="anchored" data-anchor-id="first-steps-in-the-unix-shell"><span class="header-section-number">5.2</span> First steps in the Unix shell</h3>
<section id="the-prompt" class="level4">
<h4 class="anchored" data-anchor-id="the-prompt">The prompt</h4>
<p>Inside your terminal, the “prompt” indicates that the shell is ready for a command. Our prompt at OSC should show the following pieces of information like so:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">[</span><span class="op">&lt;</span>username<span class="op">&gt;</span>@<span class="op">&lt;</span>node-name<span class="op">&gt;</span> <span class="op">&lt;</span>working-dir<span class="op">&gt;</span>]$</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>For example:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">[jelmer@p0080</span> jelmer]$ </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>We type our commands <strong>after the dollar sign</strong> <code>$</code>, and then press <kbd>Enter</kbd> to execute the command. When the command has finished executing, we’ll get our prompt back and can type a new command.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-5-contents" aria-controls="callout-5" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
How shell code is shown on this website (click to expand)
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-5" class="callout-5-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>The gray boxes like the ones shown above will be used to show the command line expressions that you should type.</p>
<p>In upcoming boxes, the prompt itself (<code>[...]$</code>) will not be shown, but only the command line expressions that you type. This is to save space and to allow you for copy-and-pasting (but I recommend typing!).</p>
<p>Paler gray boxes below, with italic text, are intended to show the <em>output</em> of commands.</p>
</div>
</div>
</div>
</section>
<section id="a-few-simple-commands-date-whoami-pwd" class="level4">
<h4 class="anchored" data-anchor-id="a-few-simple-commands-date-whoami-pwd">A few simple commands: <code>date</code>, <code>whoami</code>, <code>pwd</code></h4>
<p>The Unix shell comes with hundreds of commands. Let’s start with a few simple ones.</p>
<ul>
<li>The <code>date</code> command prints the current date and time:</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">date</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>Fri Jan 26 14:31:51 EST 2024</code></pre>
<ul>
<li>The <code>whoami</code> (who-am-i) command prints your username:</li>
</ul>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">whoami</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>jelmer</code></pre>
<ul>
<li>The <code>pwd</code> (Print Working Directory) command prints the path to the directory (=folder) you are currently located in:</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="bu">pwd</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>/fs/ess/PAS2250/ENT6703/jelmer</code></pre>
<p>All 3 of those commands provided us with some output. That output was <strong>printed to screen</strong>, which is the default behavior for nearly every Unix command.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Working directory and paths
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>When working in a Unix shell, you are always “in” a specific directory (“dir” for short) and this is called your <strong>working directory</strong>.</li>
<li>In a <strong>path</strong> (= specification of the location of a file or directory) such as that output by <code>pwd</code>, directories are separated by forward slashes <code>/</code>.</li>
<li>A <strong>leading forward slash</strong> in a path indicates the root directory of the computer.</li>
</ul>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
General shell and supercomputer tips
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Everything in the shell is case-sensitive</li>
<li><strong>Avoid spaces</strong> in file and dir names! Use e.g.&nbsp;underscores or capitalization to distinguish words.</li>
</ul>
</div>
</div>
<p><br></p>
</section>
</section>
<section id="cd-and-command-actions-defaults-and-arguments" class="level3" data-number="5.3">
<h3 data-number="5.3" class="anchored" data-anchor-id="cd-and-command-actions-defaults-and-arguments"><span class="header-section-number">5.3</span> <code>cd</code> and command actions, defaults, and arguments</h3>
<p>In the above three command line expressions:</p>
<ul>
<li>We merely typed a command and nothing else</li>
<li>The command provided some information, which was printed to screen</li>
</ul>
<p>But many commands perform an <strong>action</strong> other than providing information. For example, you can use the command <strong><code>cd</code></strong> to Change Directory (i.e.&nbsp;change your working dir). And like many commands that perform and action, it normally has no output at all.</p>
<p>Let’s use <code>cd</code> to move to another directory by specifying the <em>path</em> to that directory after the <code>cd</code> command (make sure to leave a space after <code>cd</code>!):</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> /fs/ess/PAS2250/ENT6703/share/data/fastq</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="bu">pwd</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>/fs/ess/PAS2250/ENT6703/share/data/fastq</code></pre>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
I will demonstrate Tab completion!
</div>
</div>
<div class="callout-body-container callout-body">

</div>
</div>
<p>In more abstract terms, what we did above was to provide <code>cd</code> with an <strong>argument</strong>, namely the path of the dir to move to. Arguments generally tell commands what file or directory to operate on, and come at the end of a command line expression.</p>
<p>As we’ve seen, then, <code>cd</code> gives no output when it succesfully changed the working directory. But let’s also see what happens when it does not succeed — it gives the following error:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> /fs/Ess/PAS0471</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>bash: cd: /fs/Ess/PAS0471: No such file or directory</code></pre>
<details>
<summary>
Your Turn: What was the problem with the path we specified?
</summary>
<p>We used a capital E in <code>/Ess/</code> — this should have been <code>/ess/</code>.</p>
As pointed out above, paths (dir and file specifications) are <strong>case-sensitive</strong> on Unix systems!
</details>
</section>
<section id="ls-and-command-options" class="level3" data-number="5.4">
<h3 data-number="5.4" class="anchored" data-anchor-id="ls-and-command-options"><span class="header-section-number">5.4</span> <code>ls</code> and command <em>options</em></h3>
<section id="the-default-behavior-of-ls" class="level4">
<h4 class="anchored" data-anchor-id="the-default-behavior-of-ls">The default behavior of <code>ls</code></h4>
<p>The <code>ls</code> command, short for “list”, is a command to list files and directories:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>data  metadata  README.md</code></pre>
<p><em>(You should still be in <code>/fs/ess/PAS2250/ENT6703/share</code>. If not, <code>cd</code> there first.)</em></p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-9-contents" aria-controls="callout-9" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<code>ls</code> output colors (click to expand)
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-9" class="callout-9-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Unfortunately, the <code>ls</code> output shown above does not show the different colors you should see in your shell — here are some of the most common ones:</p>
<ul>
<li>Entries in <span style="color: #0328fc">blue</span> are directories (like <code>data</code> and <code>metadata</code> above)</li>
<li>Entries in black are regular files (like <code>README.md</code> above)</li>
<li>Entries in <span style="color: #d92118">red</span> are compressed files (we’ll see an example soon).</li>
</ul>
</div>
</div>
</div>
<p>The default way that <code>ls</code> shows the output (e.g., non-recursively, all on one line, no details) can be changed by providing <code>ls</code> with <strong>options</strong> and/or arguments.</p>
</section>
<section id="options-to-ls" class="level4">
<h4 class="anchored" data-anchor-id="options-to-ls">Options (to <code>ls</code>)</h4>
<p>In general, whereas arguments tell a command <em>what to operate on</em>, options will <em>modify its behavior</em>. For example, we can call <code>ls</code> with the <strong>option <code>-l</code></strong> (a dash followed by a lowercase L):</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-l</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>total 17
drwxr-xr-x 3 jelmer PAS0471 4096 Jul 27 11:53 data
drwxr-xr-x 2 jelmer PAS0471 4096 Jul 27 11:54 metadata
-rw-r--r-- 1 jelmer PAS0471  963 Jul 27 16:48 README.md</code></pre>
<p>Notice that it lists the same three items as our first <code>ls</code> call above, but <strong>printed in a different format</strong>: one item per line, with additional information included, such as the date and time each file was last modified, and the file sizes (to the left of the date) in bytes.</p>
<p>Let’s add another option, <code>-h</code>:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-l</span> <span class="at">-h</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>total 17K
drwxr-xr-x 3 jelmer PAS0471 4.0K Jul 27 11:53 data
drwxr-xr-x 2 jelmer PAS0471 4.0K Jul 27 11:54 metadata
-rw-r--r-- 1 jelmer PAS0471  964 Jul 27 17:48 README.md</code></pre>
<details>
<summary>
Your Turn: What is different about the output, and what does that mean?
</summary>
<p>The difference is in the format of the column reporting the sizes of the items listed.</p>
We now have “Human-readable filesizes” (hence <code>-h</code>), where sizes on the scale of kilobytes will be shown in <code>K</code>s, of megabytes in <code>M</code>s, and of gigabytes in <code>G</code>s.
</details>
<p>Conveniently, options can be pasted together as follows:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-lh</span>   <span class="co"># Output not shown, same as above</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="combining-options-and-arguments" class="level4">
<h4 class="anchored" data-anchor-id="combining-options-and-arguments">Combining options and arguments</h4>
<p>Arguments to <code>ls</code> should be <strong>dirs or files to operate on</strong>. For example, if we wanted to see what’s inside the <code>data</code> dir, instead of inside our working dir, we could type:</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>fastq</code></pre>
<p>Well, that’s not much information, just another dir — so let’s look inside that:</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> data/fastq  <span class="co"># These will be shown in red in your output, since they are compressed</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>ASPC1_A178V_R1.fastq.gz  ASPC1_G31V_R2.fastq.gz      Miapaca2_G31V_R1.fastq.gz
ASPC1_A178V_R2.fastq.gz  Miapaca2_A178V_R1.fastq.gz  Miapaca2_G31V_R2.fastq.gz
ASPC1_G31V_R1.fastq.gz   Miapaca2_A178V_R2.fastq.gz</code></pre>
<p>Ah, gzipped FASTQ files! These contain our sequence data, and we’ll go and explore them in a bit.</p>
<p>Finally, we can <strong>combine options and arguments</strong>, and let’s do so take a closer look at our dir with FASTQ files — now the <code>-h</code> option is especially useful because it makes it easy to see that the files vary between 4.1 MB and 5.3 MB in size:</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-lh</span> data/fastq</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>total 38M
-rw-r--r-- 1 jelmer PAS0471 4.1M Jul 27 11:53 ASPC1_A178V_R1.fastq.gz
-rw-r--r-- 1 jelmer PAS0471 4.2M Jul 27 11:53 ASPC1_A178V_R2.fastq.gz
-rw-r--r-- 1 jelmer PAS0471 4.1M Jul 27 11:53 ASPC1_G31V_R1.fastq.gz
-rw-r--r-- 1 jelmer PAS0471 4.3M Jul 27 11:53 ASPC1_G31V_R2.fastq.gz
-rw-r--r-- 1 jelmer PAS0471 5.1M Jul 27 11:53 Miapaca2_A178V_R1.fastq.gz
-rw-r--r-- 1 jelmer PAS0471 5.3M Jul 27 11:53 Miapaca2_A178V_R2.fastq.gz
-rw-r--r-- 1 jelmer PAS0471 5.1M Jul 27 11:53 Miapaca2_G31V_R1.fastq.gz
-rw-r--r-- 1 jelmer PAS0471 5.3M Jul 27 11:53 Miapaca2_G31V_R2.fastq.gz</code></pre>
</section>
</section>
<section id="unix-shell-recap" class="level3" data-number="5.5">
<h3 data-number="5.5" class="anchored" data-anchor-id="unix-shell-recap"><span class="header-section-number">5.5</span> Unix shell recap</h3>
<p>We’ve learned about structure of command line expressions in the Unix shell, which include: the <em>command</em> itself, <em>options</em>, <em>arguments</em>, and <em>output</em> (including, in some cases, error messages).</p>
<p>A few general points to remember are that:</p>
<ul>
<li>Commands that take <strong>actions</strong> like changing directory will by default <strong>not print output</strong>.</li>
<li>Using <strong>options</strong> (<code>ls -l</code>), we can modify the behavior of a command, and using <strong>arguments</strong> (<code>ls data</code>), we can modify what it operates on in the first place.</li>
</ul>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
General shell tips
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li><p>If you hit the <kbd>⇧</kbd> (up arrow) once, you’ll retrieve your most recent command, and if you keep hitting it, you’ll go further back. The<kbd>⇩</kbd> (down arrow) will go the other way: towards the present.</p></li>
<li><p>To execute a command, <strong>you can press <kbd>Enter</kbd> regardless of where your cursor is</strong> on the line; it does <em>not</em> need to be at the end of the line.</p></li>
<li><p>If your prompt is missing, the shell is either still busy executing your command, or you typed an incomplete command. To abort in either of these two scenarios, press <kbd>Ctrl</kbd>+<kbd>C</kbd> and you’ll get your prompt back.</p></li>
</ul>
</div>
</div>
<details>
<summary>
On your own: practice with the above tips (click to expand)
</summary>
<p>A somewhat silly example is the sleep command, which you can use to make the computer wait between successive commands:</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># This will "run" for 60 seconds, after which you get your prompt back </span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sleep</span> 60s</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Press Ctrl + C to get your prompt back sooner!</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<p><br></p>
</section>
</section>
<section id="fastq-files" class="level2" data-number="6">
<h2 data-number="6" class="anchored" data-anchor-id="fastq-files"><span class="header-section-number">6</span> FASTQ files</h2>
<section id="recap-the-fastq-format" class="level4">
<h4 class="anchored" data-anchor-id="recap-the-fastq-format">Recap: The FASTQ format</h4>
<p>FASTQ is the most common HTS read data file format, and like most genomic data files, these are plain text files. Each sequence that is read by the sequencer (i.e., each “read”) forms <strong>one FASTQ entry represented by four lines</strong>. The lines contain, respectively:</p>
<ol type="1">
<li>A <strong>header</strong> that starts with <code>@</code> and e.g.&nbsp;uniquely identifies the read</li>
<li>The <strong>sequence</strong> itself</li>
<li>A <strong><code>+</code></strong> (plus sign)</li>
<li>One-character <strong>quality scores</strong> for each base in the sequence</li>
</ol>
<figure class="figure">
<p align="center">
<img src="../lectures/img_sequencing/fastq_header.png" width="85%" class="figure-img">
</p><figcaption class="figure-caption">
One entry (read) in a FASTQ file covers 4 lines. <br>The header line is annotated, with some of the more useful components highlighted in red. <br>For viewing purposes, this read (at only 56 bp) is shorter than regular Illumina read lengths.
</figcaption>
<p></p>
</figure>
</section>
<section id="getting-your-own-copy" class="level4">
<h4 class="anchored" data-anchor-id="getting-your-own-copy">Getting your own copy</h4>
<p>To get you your own copy of the FASTQ files, we’ll use the Unix copy command <code>cp</code> as follows:</p>
<ul>
<li>Option <code>-r</code> will enable “<strong>r</strong>ecursive” (=dirs, not just files) copying</li>
<li>Option <code>-v</code> will turn on “<strong>v</strong>erbose” output: it will report what it’s copying</li>
<li>The first argument is the source directory</li>
<li>The second argument is the target directory, with <code>.</code> being shorthand for the current working dir</li>
</ul>
<div class="sourceCode" id="cb28"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cp</span> <span class="at">-rv</span> /fs/scratch/PAS2250/ENT6703/share/data .</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>‘/fs/ess/PAS2250/ENT6703/demo/202307_rnaseq/data/fastq’ -&gt; ‘data/fastq’
‘/fs/ess/PAS2250/ENT6703/demo/202307_rnaseq/data/fastq/Miapaca2_A178V_R1.fastq.gz’ -&gt; ‘data/fastq/Miapaca2_A178V_R1.fastq.gz’
‘/fs/ess/PAS2250/ENT6703/demo/202307_rnaseq/data/fastq/ASPC1_G31V_R2.fastq.gz’ -&gt; ‘data/fastq/ASPC1_G31V_R2.fastq.gz’
‘/fs/ess/PAS2250/ENT6703/demo/202307_rnaseq/data/fastq/ASPC1_A178V_R2.fastq.gz’ -&gt; ‘data/fastq/ASPC1_A178V_R2.fastq.gz’</code></pre>
<p>The FASTQ files all have a <code>.gz</code> extension (and should listed in red in your terminal), indicating they are “gzip-compressed”. This is a common type of compression for large genomic files.</p>
</section>
<section id="interlude-viewing-the-metadata-file" class="level4">
<h4 class="anchored" data-anchor-id="interlude-viewing-the-metadata-file">Interlude: Viewing the metadata file</h4>
<p>You can do so by finding and clicking on them in the Explorer in the side bar.</p>
<p>[TBA]</p>
</section>
<section id="viewing-the-fastq-files" class="level4">
<h4 class="anchored" data-anchor-id="viewing-the-fastq-files">Viewing the FASTQ files</h4>
<p>While we can easily open small to medium-size files in the editor pane, “visual editors” like that not work as well for very large files.</p>
<p>A handy command to view text files of any size is <strong><code>less</code></strong>, which opens them up in a “pager” within your shell. That is, you will not get your prompt back until you press <strong><kbd>q</kbd> to quit</strong> <code>less</code>, and you can e.g.&nbsp;scroll/move around in the file.</p>
<p>Try it with one of the FASTQ files:</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">less</span> data/fastq/ASPC1_A178V_R1.fastq.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Besides scrolling with your mouse, its easiest to move around with up and down arrows and, if you have them, <kbd>PgUp</kbd> and <kbd>PgDn</kbd> (also, <kbd>u</kbd> will move up half a page and <kbd>d</kbd> will move down half a page).</p>
<p>If you find yourself scrolling down and down to try and reach the end of the file, you can instead press <kbd>G</kbd> to go to the very end right away (and <kbd>g</kbd> to go back to the top).</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Avoid line-wrapping by <code>less</code>
</div>
</div>
<div class="callout-body-container callout-body">
<p>Depending on your zoom level and the length of reads in your FASTQ file, some lines may contain too many characters to fit on your screen. If that’s the case, <code>less</code> will by default “wrap” those lines onto the next line on your screen, so characters won’t run off the screen on the right-hand side. That may be useful when the file contains text you’re trying to read in full, but it is often confusing for files like FASTQ as well as for tabular files.</p>
<p>To turn off line-wrapping, call <code>less</code> with the <code>-S</code> option:</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">less</span> <span class="at">-S</span> data/fastq/ASPC1_A178V_R1.fastq.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<p><br></p>
</section>
</section>
<section id="running-fastqc" class="level2" data-number="7">
<h2 data-number="7" class="anchored" data-anchor-id="running-fastqc"><span class="header-section-number">7</span> Running FastQC</h2>
<section id="what-is-fastqc" class="level3" data-number="7.1">
<h3 data-number="7.1" class="anchored" data-anchor-id="what-is-fastqc"><span class="header-section-number">7.1</span> What is FastQC?</h3>
<p>A useful example of a genomics tool with a CLI is FastQC, for <strong>quality control of FASTQ files</strong>. It is ubiquitous because nearly all high-throughput sequencing data comes in FASTQ files, and your first step is always to check the quality of the reads.</p>
<p>FastQC produces visualizations and assessments of aspects of your reads such as adapter content, and, as shown below, <strong>mean base quality along the read</strong>:</p>
<div class="columns">
<div class="column" style="width:50%;">
<p><img src="img_HTS/fastqc_good.png" width="85%"></p>
</div><div class="column" style="width:50%;">
<p><img src="img_HTS/fastqc_bad.png" width="85%"></p>
</div>
</div>
</section>
<section id="running-fastqc-1" class="level3" data-number="7.2">
<h3 data-number="7.2" class="anchored" data-anchor-id="running-fastqc-1"><span class="header-section-number">7.2</span> Running FastQC</h3>
<p>To run FastQC, you use the command <code>fastqc</code>.</p>
<p>Command-line programs are typically run non-interactively, so we don’t fire up the program first, and tell it what to do as we go along. Instead, we at once issue a complete set of instructions for the program to do what we would like it to.</p>
<p>If we want to analyze one of our FASTQ files with default FastQC settings, a complete FastQC command to do so would simply be <code>fastqc</code> followed by the name of the file (like with, say, <code>ls</code>!):</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="ex">fastqc</span> data/fastq/XX.fastq.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>fastqc: command not found</code></pre>
<p>However, there is one wrinkle, as you can see above. It turns out that a FastQC installation <em>is</em> already available to us at OSC <a href="#fn8" class="footnote-ref" id="fnref8" role="doc-noteref"><sup>8</sup></a>, but we do have to <strong>load it</strong> before we can use it. We can do so as follows:</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load fastqc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Now, let’s try again:</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="ex">fastqc</span> /fs/scratch/PAS2250/ENT6703/data/sample1.fastq.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>#&gt; Started analysis of sample1.fastq.gz
#&gt; Approx 5% complete for sample1.fastq.gz
#&gt; Approx 10% complete for sample1.fastq.gz
#&gt; Approx 15% complete for sample1.fastq.gz
#&gt; [truncated]</code></pre>
<p>Success!</p>
<p><br></p>
</section>
</section>
<section id="interpreting-fastqcs-output" class="level2" data-number="8">
<h2 data-number="8" class="anchored" data-anchor-id="interpreting-fastqcs-output"><span class="header-section-number">8</span> Interpreting FastQC’s output</h2>
<p><br></p>
</section>
<section id="next-steps" class="level2" data-number="9">
<h2 data-number="9" class="anchored" data-anchor-id="next-steps"><span class="header-section-number">9</span> Next steps</h2>
<section id="next-steps-in-an-rna-seq-workflow" class="level3" data-number="9.1">
<h3 data-number="9.1" class="anchored" data-anchor-id="next-steps-in-an-rna-seq-workflow"><span class="header-section-number">9.1</span> Next steps in an RNA-seq workflow</h3>
</section>
<section id="next-steps-to-run-the-analyses-more-efficiently" class="level3" data-number="9.2">
<h3 data-number="9.2" class="anchored" data-anchor-id="next-steps-to-run-the-analyses-more-efficiently"><span class="header-section-number">9.2</span> Next steps to run the analyses more efficiently</h3>
<p>I’ve shown you the main pieces of the computational infrastructure for<br>
“<strong>command-line genomics</strong>”. We’ve seen a very basic example of loading and running a command-line tool at OSC.</p>
<p>The missing pieces for a fuller example of how such tools are run in the context of an actual genomics project are (if we stay with FastQC):</p>
<ul>
<li><p>Putting the command to run FastQC in a <strong>“shell script”</strong>.<br>
<em>(This used the same language (Bash) as the commands you’d type interactively,</em> <em>so at its most basic this involves pasting those commands into a text file.)</em></p></li>
<li><p><strong>Submitting the script</strong> to the SLURM scheduler queue as a “batch job”.<br>
<em>(At its most basic, this involves putting <code>sbatch</code> in front of the script name.)</em></p></li>
<li><p>To make use of the capabilities of the supercomputer and speeding up our analysis, we can <strong>submit multiple jobs <em>in parallel</em> using a loop</strong>.</p></li>
</ul>
<p><br></p>
<hr>
<p><br></p>
</section>
</section>
<section id="to-learn-more" class="level2" data-number="10">
<h2 data-number="10" class="anchored" data-anchor-id="to-learn-more"><span class="header-section-number">10</span> To learn more</h2>
<section id="overview-of-computational-skills-for-command-line-genomics" class="level4">
<h4 class="anchored" data-anchor-id="overview-of-computational-skills-for-command-line-genomics">Overview of computational skills for “command-line genomics”</h4>
<ul>
<li><strong>The core skills</strong>:
<ul>
<li>Unix shell basics – the commonly used commands</li>
<li>Some shell <em>scripting</em> basics</li>
<li>SLURM basics to submit and manage your batch jobs</li>
<li>R for “downstream”, statistical and visualization tasks</li>
</ul></li>
<li><strong>When you start doing genomics projects more often</strong>:
<ul>
<li>Using <code>conda</code> or containers for software</li>
<li>Unix data tools (<code>grep</code>, <code>sed</code>, <code>awk</code>, etc)</li>
</ul></li>
<li><strong>When you want to become proficient in applied bioinformatics</strong>:
<ul>
<li>Version control with <code>git</code></li>
<li>More advanced: formal workflow/pipeline management tools (e.g.&nbsp;<code>Nextflow</code>)</li>
<li>More advanced: Python (or advanced R) for custom data processing</li>
</ul></li>
</ul>
</section>
<section id="resources-for-further-learning" class="level4">
<h4 class="anchored" data-anchor-id="resources-for-further-learning">Resources for further learning</h4>
<ul>
<li><strong>OSC</strong>
<ul>
<li>https://mcic-osu.github.io/rnaseq-intro/modules/A01_osc.html</li>
<li><a href="https://www.osc.edu/supercomputing/training">OSC’s online asynchronous courses</a></li>
<li><a href="https://www.osc.edu/resources/getting_started/new_user_resource_guide">OSC’s new User Resource Guide</a></li>
</ul></li>
<li><strong>VS Code</strong>
<ul>
<li><a href="https://mcic-osu.github.io/rnaseq-intro/modules/A02_vscode.html" class="uri">https://mcic-osu.github.io/rnaseq-intro/modules/A02_vscode.html</a></li>
</ul></li>
<li><strong>Unix shell</strong>
<ul>
<li><a href="https://www.osc.edu/documentation/tutorials/unix_basics">OSC’s UNIX Basics</a></li>
<li><a href="https://www.learnenough.com/command-line-tutorial" class="uri">https://www.learnenough.com/command-line-tutorial</a></li>
<li><a href="https://cvw.cac.cornell.edu/Linux/" class="uri">https://cvw.cac.cornell.edu/Linux/</a></li>
</ul></li>
<li><strong>OSU courses and workshops</strong>
<ul>
<li>Genome Analytics course (<code>HCS 7004</code>)</li>
<li>Microbiome Informatics course (<code>MICRBIO 8161</code>)</li>
<li>The online materials for the workshop “<a href="https://mcic-osu.github.io/cl-workshop-22">Command line basics for genomic analysis at OSC</a>” (Mike Sovic &amp; Jelmer Poelstra, August 2022)</li>
<li>The online materials for the course <a href="https://mcic-osu.github.io/pracs-sp21/">“Practical Computing Skills for Biologists”</a> (Jelmer Poelstra, Spring 2021)</li>
</ul></li>
<li><strong>Books</strong>
<ul>
<li><a href="https://open.oregonstate.education/computationalbiology">A Primer for Computational Biology</a> (Shawn T. O’ Neil, 2019) <em>(available online!)</em></li>
<li>Computing Skills for Biologists: A Toolbox (Wilmes &amp; Allesino, 2019)</li>
<li>Bioinformatics Data Skills (Vince Buffalo, 2015)</li>
<li>The Linux Command Line (William Shotts, 2019)</li>
</ul></li>
</ul>


</section>
</section>


<a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a><div id="quarto-appendix" class="default"><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><h2 class="anchored quarto-appendix-heading">Footnotes</h2>

<ol>
<li id="fn1"><p>Cloud computing is an alternative, but won’t be covered here.<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn2"><p>You can be associated with multiple projects and for each one, a scratch and a project directory is added.<a href="#fnref2" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn3"><p>Though this is not meant for large (&gt;1 GB) transfers. Different methods are available for those but are outside the scope of this introductions.<a href="#fnref3" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn4"><p>The menu item next to that, <strong>My Interactive Sessions</strong>, will list the currently active and finished Interactive App sessions.<a href="#fnref4" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn5"><p>we’ll be kicked off as soon as that amount of time has passed!<a href="#fnref5" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn6"><p>If you’ve closed the <code>Welcome</code> document but want it back, click<br>
<i class="fa-solid fa-bars" aria-label="bars"></i> &nbsp; =&gt; &nbsp; <code>Help</code> &nbsp; =&gt; &nbsp; <code>Welcome</code>.<a href="#fnref6" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn7"><p>And the most common Unix shell is the <strong>Bash shell</strong>, which runs the Bash language.<a href="#fnref7" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn8"><p>full list of installed software: <a href="https://www.osc.edu/resources/available_software/software_list" class="uri">https://www.osc.edu/resources/available_software/software_list</a><a href="#fnref8" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>